// @xmtp/agent-sdk@2.2.0/dist/index.js
import { Group, Dm, Client, encryptAttachment, decryptAttachment, isMarkdown, isText, isReply, isReaction, isReadReceipt, isRemoteAttachment, isTransactionReference, isWalletSendCalls, encodeMarkdown, encodeText, isHexString, ApiUrls, DecodedMessage, isMultiRemoteAttachment, isLeaveRequest, isGroupUpdated, isIntent, isAttachment, isActions } from '@xmtp/node-sdk';
export * from '@xmtp/node-sdk';
import EventEmitter from 'node:events';
import fs from 'node:fs';
import path from 'node:path';
import { contentTypesAreEqual } from '@xmtp/content-type-primitives';
import { toBytes, createWalletClient, http, isAddress, encodeFunctionData, toHex, createPublicClient } from 'viem';
import { generatePrivateKey, privateKeyToAccount } from 'viem/accounts';
import { sepolia } from 'viem/chains';
import { monitorEventLoopDelay, performance } from 'node:perf_hooks';
import v8 from 'node:v8';
import { escape } from 'node:querystring';

const version = "2.2.0";

const fromSelf = (message, client) => {
    return message.senderInboxId === client.inboxId;
};
const hasContent = (message) => {
    return message.content !== undefined && message.content !== null;
};
const isDM = (conversation) => {
    return conversation instanceof Dm;
};
const isGroup = (conversation) => {
    return conversation instanceof Group;
};
const isGroupAdmin = (conversation, message) => {
    if (isGroup(conversation)) {
        return conversation.isAdmin(message.senderInboxId);
    }
    return false;
};
const isGroupSuperAdmin = (conversation, message) => {
    if (isGroup(conversation)) {
        return conversation.isSuperAdmin(message.senderInboxId);
    }
    return false;
};
const usesCodec = (message, codecClass) => {
    return contentTypesAreEqual(message.contentType, new codecClass().contentType);
};
const filter = {
    fromSelf,
    hasContent,
    isDM,
    isGroup,
    isGroupAdmin,
    isGroupSuperAdmin,
    usesCodec,
};
const f = filter;

const validLogLevels = [
    "Off" /* LogLevel.Off */,
    "Error" /* LogLevel.Error */,
    "Warn" /* LogLevel.Warn */,
    "Info" /* LogLevel.Info */,
    "Debug" /* LogLevel.Debug */,
    "Trace" /* LogLevel.Trace */,
];
const isLogLevel = (level) => {
    return validLogLevels.includes(level);
};
const getValidLogLevels = () => {
    return [...validLogLevels];
};
const parseLogLevel = (rawLevel) => {
    const normalizedLevel = rawLevel.charAt(0).toUpperCase() + rawLevel.slice(1).toLowerCase();
    if (isLogLevel(normalizedLevel)) {
        return normalizedLevel;
    }
    return null;
};
const logDetails = async (agent) => {
    const xmtp = `\x1b[38;2;252;76;52m
    ██╗  ██╗███╗   ███╗████████╗██████╗ 
    ╚██╗██╔╝████╗ ████║╚══██╔══╝██╔══██╗
     ╚███╔╝ ██╔████╔██║   ██║   ██████╔╝
     ██╔██╗ ██║╚██╔╝██║   ██║   ██╔═══╝ 
    ██╔╝ ██╗██║ ╚═╝ ██║   ██║   ██║     
    ╚═╝  ╚═╝╚═╝     ╚═╝   ╚═╝   ╚═╝     
  \x1b[0m`;
    const client = agent.client;
    const clientsByAddress = client.accountIdentifier?.identifier;
    const inboxId = client.inboxId;
    const installationId = client.installationId;
    const env = client.options?.env ?? "dev";
    const urls = [`http://xmtp.chat/${env}/dm/${clientsByAddress}`];
    const conversations = await client.conversations.list();
    const inboxState = await client.preferences.inboxState();
    const keyPackageStatuses = await client.fetchKeyPackageStatuses([
        installationId,
    ]);
    let createdDate = new Date();
    let expiryDate = new Date();
    // Extract key package status for the specific installation
    const keyPackageStatus = keyPackageStatuses[installationId] ?? {};
    if (keyPackageStatus.lifetime) {
        createdDate = new Date(Number(keyPackageStatus.lifetime.notBefore) * 1000);
        expiryDate = new Date(Number(keyPackageStatus.lifetime.notAfter) * 1000);
    }
    console.log(`
    ${xmtp}
    
    ✓ XMTP Client:
    • InboxId: ${inboxId}
    • LibXMTP Version: ${agent.libxmtpVersion}
    • Address: ${clientsByAddress}
    • Conversations: ${conversations.length}
    • Installations: ${inboxState.installations.length}
    • InstallationId: ${installationId}
    • Key Package created: ${createdDate.toLocaleString()}
    • Key Package valid until: ${expiryDate.toLocaleString()}
    • Networks: ${env}
    ${urls.map((url) => `• URL: ${url}`).join("\n")}`);
};
/**
 * Returns a URL to test your agent on https://xmtp.chat/ (for development purposes only).
 *
 * @param client - Your XMTP client
 * @returns The URL to test your agent with
 */
const getTestUrl = (client) => {
    const address = client.accountIdentifier?.identifier;
    const env = client.options?.env ?? "dev";
    return `http://xmtp.chat/${env}/dm/${address}`;
};
const getInstallationInfo = async (client) => {
    const myInboxId = client.inboxId;
    const myInstallationId = client.installationId;
    const inboxStates = await Client.fetchInboxStates([myInboxId], client.options?.env, client.options?.gatewayHost);
    const installations = inboxStates.find((state) => state.inboxId === myInboxId)?.installations ||
        [];
    const sortedInstallations = [...installations].sort((a, b) => {
        const aTime = a.clientTimestampNs ?? 0n;
        const bTime = b.clientTimestampNs ?? 0n;
        return bTime > aTime ? 1 : bTime < aTime ? -1 : 0;
    });
    const mostRecentInstallation = sortedInstallations[0];
    const myInstallationIdHex = Buffer.from(client.installationIdBytes).toString("hex");
    const info = {
        totalInstallations: installations.length,
        installationId: myInstallationId,
        mostRecentInstallationId: null,
        isMostRecent: false,
    };
    if (mostRecentInstallation) {
        const mostRecentIdHex = Buffer.from(mostRecentInstallation.bytes).toString("hex");
        info.mostRecentInstallationId = mostRecentIdHex;
        info.isMostRecent = myInstallationIdHex === mostRecentIdHex;
    }
    return info;
};

const createUser = (key, chain = sepolia) => {
    const accountKey = key ?? generatePrivateKey();
    const account = privateKeyToAccount(accountKey);
    return {
        key: accountKey,
        account,
        wallet: createWalletClient({
            account,
            chain,
            transport: http(),
        }),
    };
};
const createIdentifier = (user) => ({
    identifier: user.account.address.toLowerCase(),
    identifierKind: 0 /* IdentifierKind.Ethereum */,
});
const createSigner = (user) => {
    const identifier = createIdentifier(user);
    return {
        type: "EOA",
        getIdentifier: () => identifier,
        signMessage: async (message) => {
            const signature = await user.wallet.signMessage({
                account: user.account,
                message,
            });
            return toBytes(signature);
        },
    };
};

class AgentError extends Error {
    #code;
    constructor(code, message, cause) {
        super(message, { cause });
        this.#code = code;
    }
    get code() {
        return this.#code;
    }
}
class AgentStreamingError extends AgentError {
}

class ClientContext {
    #client;
    constructor({ client }) {
        this.#client = client;
    }
    getClientAddress() {
        return this.#client.accountIdentifier?.identifier;
    }
    get client() {
        return this.#client;
    }
}

/**
 * Downloads and decrypts a remote attachment.
 *
 * @param remoteAttachment - The remote attachment metadata containing the downloadd URL and encryption keys
 * @param agent - The agent instance used to lookup the necessary decoding codec
 * @returns A promise that resolves with the decrypted attachment
 */
async function downloadRemoteAttachment(remoteAttachment) {
    const response = await fetch(remoteAttachment.url);
    if (!response.ok) {
        throw new Error(`unable to fetch remote attachment at "${remoteAttachment.url}": [${response.status}] ${response.statusText}`);
    }
    const payload = new Uint8Array(await response.arrayBuffer());
    return decryptAttachment(payload, remoteAttachment);
}
/**
 * Creates a remote attachment object from an encrypted attachment and file URL.
 *
 * @param encryptedAttachment - The encrypted attachment containing encryption keys and metadata
 * @param fileUrl - The URL where the encrypted attachment can be downloaded
 * @returns A remote attachment object with all necessary metadata for retrieval and decryption
 */
function createRemoteAttachment(encryptedAttachment, fileUrl) {
    const url = new URL(fileUrl);
    return {
        contentDigest: encryptedAttachment.contentDigest,
        contentLength: encryptedAttachment.payload.length,
        filename: encryptedAttachment.filename,
        nonce: encryptedAttachment.nonce,
        salt: encryptedAttachment.salt,
        scheme: url.protocol,
        secret: encryptedAttachment.secret,
        url: url.toString(),
    };
}
/**
 * Creates a remote attachment from a file by encrypting it and uploading it to a remote storage.
 * This is a convenience function that combines file processing, encryption, uploading, and
 * remote attachment creation into a single operation.
 *
 * @param unencryptedFile - The unencrypted file to process and upload
 * @param uploadCallback - A callback function that receives the encrypted attachment and returns the URL where it was uploaded
 * @returns A promise that resolves with a remote attachment containing all necessary metadata for retrieval and decryption
 */
async function createRemoteAttachmentFromFile(unencryptedFile, uploadCallback) {
    const arrayBuffer = await unencryptedFile.arrayBuffer();
    const attachment = new Uint8Array(arrayBuffer);
    const attachmentData = {
        content: attachment,
        filename: unencryptedFile.name,
        mimeType: unencryptedFile.type,
    };
    const encryptedAttachment = encryptAttachment(attachmentData);
    const fileUrl = await uploadCallback(encryptedAttachment);
    return createRemoteAttachment(encryptedAttachment, fileUrl);
}

class ConversationContext extends ClientContext {
    #conversation;
    constructor({ conversation, client, }) {
        super({ client });
        this.#conversation = conversation;
    }
    isDm() {
        return filter.isDM(this.#conversation);
    }
    isGroup() {
        return filter.isGroup(this.#conversation);
    }
    async sendRemoteAttachment(unencryptedFile, uploadCallback) {
        const remoteAttachment = await createRemoteAttachmentFromFile(unencryptedFile, uploadCallback);
        await this.#conversation.sendRemoteAttachment(remoteAttachment);
    }
    get conversation() {
        return this.#conversation;
    }
    get isAllowed() {
        return this.#conversation.consentState() === 1 /* ConsentState.Allowed */;
    }
    get isDenied() {
        return this.#conversation.consentState() === 2 /* ConsentState.Denied */;
    }
    get isUnknown() {
        return this.#conversation.consentState() === 0 /* ConsentState.Unknown */;
    }
}

class MessageContext extends ConversationContext {
    #message;
    constructor({ message, conversation, client, }) {
        super({ conversation, client });
        this.#message = message;
    }
    usesCodec(codecClass) {
        return filter.usesCodec(this.#message, codecClass);
    }
    isMarkdown() {
        return isMarkdown(this.#message);
    }
    isText() {
        return isText(this.#message);
    }
    isReply() {
        return isReply(this.#message);
    }
    isReaction() {
        return isReaction(this.#message);
    }
    isReadReceipt() {
        return isReadReceipt(this.#message);
    }
    isRemoteAttachment() {
        return isRemoteAttachment(this.#message);
    }
    isTransactionReference() {
        return isTransactionReference(this.#message);
    }
    isWalletSendCalls() {
        return isWalletSendCalls(this.#message);
    }
    async sendReaction(content, schema = 1 /* ReactionSchema.Unicode */) {
        const reaction = {
            action: 1 /* ReactionAction.Added */,
            reference: this.#message.id,
            referenceInboxId: this.#message.senderInboxId,
            schema,
            content,
        };
        await this.conversation.sendReaction(reaction);
    }
    async #sendReply(content) {
        await this.conversation.sendReply({
            content,
            reference: this.#message.id,
            referenceInboxId: this.#message.senderInboxId,
        });
    }
    async sendMarkdownReply(markdown) {
        await this.#sendReply(encodeMarkdown(markdown));
    }
    async sendTextReply(text) {
        await this.#sendReply(encodeText(text));
    }
    async getSenderAddress() {
        const inboxState = await this.client.preferences.getInboxStates([
            this.#message.senderInboxId,
        ]);
        return inboxState[0]?.identifiers[0]?.identifier;
    }
    get message() {
        return this.#message;
    }
}

class Agent extends EventEmitter {
    #client;
    #conversationsStream;
    #messageStream;
    #middleware = [];
    #errorMiddleware = [];
    #errors = Object.freeze({
        use: (...errorMiddleware) => {
            for (const emw of errorMiddleware) {
                if (Array.isArray(emw)) {
                    this.#errorMiddleware.push(...emw);
                }
                else if (typeof emw === "function") {
                    this.#errorMiddleware.push(emw);
                }
            }
            return this.#errors;
        },
    });
    #defaultErrorHandler = (currentError) => {
        const emittedError = currentError instanceof Error
            ? currentError
            : new AgentError(9999, `Unhandled error caught by default error middleware.`, currentError);
        this.emit("unhandledError", emittedError);
    };
    #isLocked = false;
    constructor({ client }) {
        super();
        this.#client = client;
    }
    static async create(signer, 
    // Note: we need to omit this so that "Client.create" can correctly infer the codecs.
    options) {
        const initializedOptions = { ...(options ?? {}) };
        initializedOptions.appVersion ??= `agent-sdk/${version}`;
        initializedOptions.disableDeviceSync ??= true;
        if (process.env.XMTP_FORCE_DEBUG_LEVEL) {
            const rawLevel = process.env.XMTP_FORCE_DEBUG_LEVEL;
            const logLevel = parseLogLevel(rawLevel);
            if (logLevel) {
                initializedOptions.loggingLevel = logLevel;
            }
            else {
                console.warn(`[WARNING] Invalid XMTP_FORCE_DEBUG_LEVEL "${rawLevel}". Defaulting to "${"Warn" /* LogLevel.Warn */}". Valid values are: ${getValidLogLevels().join(", ")}`);
                initializedOptions.loggingLevel = "Warn" /* LogLevel.Warn */;
            }
            initializedOptions.structuredLogging = true;
        }
        const client = await Client.create(signer, {
            ...initializedOptions,
            codecs: initializedOptions.codecs,
        });
        const info = await getInstallationInfo(client);
        if (info.totalInstallations > 1 && info.isMostRecent) {
            console.warn(`[WARNING] You have "${info.totalInstallations}" installations. Installation ID "${info.installationId}" is the most recent. Make sure to persist and reload your installation data. If you exceed the installation limit, your Agent will stop working. Read more: https://docs.xmtp.org/agents/build-agents/local-database#installation-limits-and-revocation-rules`);
        }
        return new Agent({ client });
    }
    static async createFromEnv(
    // Note: we need to omit this so that "Client.create" can correctly infer the codecs.
    options) {
        const { XMTP_DB_DIRECTORY, XMTP_DB_ENCRYPTION_KEY, XMTP_ENV, XMTP_WALLET_KEY, XMTP_GATEWAY_HOST, } = process.env;
        if (!isHexString(XMTP_WALLET_KEY)) {
            throw new AgentError(1000, `XMTP_WALLET_KEY env is not in hex (0x) format.`);
        }
        const signer = createSigner(createUser(XMTP_WALLET_KEY));
        const initializedOptions = { ...(options ?? {}) };
        initializedOptions.dbEncryptionKey =
            typeof XMTP_DB_ENCRYPTION_KEY === "string"
                ? isHexString(XMTP_DB_ENCRYPTION_KEY)
                    ? XMTP_DB_ENCRYPTION_KEY
                    : `0x${XMTP_DB_ENCRYPTION_KEY}`
                : undefined;
        if (XMTP_ENV && Object.keys(ApiUrls).includes(XMTP_ENV)) {
            initializedOptions.env = XMTP_ENV;
        }
        if (typeof XMTP_GATEWAY_HOST === "string") {
            initializedOptions.gatewayHost = XMTP_GATEWAY_HOST;
        }
        if (typeof XMTP_DB_DIRECTORY === "string") {
            fs.mkdirSync(XMTP_DB_DIRECTORY, { recursive: true, mode: 0o700 });
            initializedOptions.dbPath = (inboxId) => {
                const dbPath = path.join(XMTP_DB_DIRECTORY, `xmtp-${inboxId}.db3`);
                console.info(`Saving local database to "${dbPath}"`);
                return dbPath;
            };
        }
        return this.create(signer, initializedOptions);
    }
    get libxmtpVersion() {
        return this.#client.libxmtpVersion;
    }
    use(...middleware) {
        for (const mw of middleware) {
            if (Array.isArray(mw)) {
                this.#middleware.push(...mw);
            }
            else if (typeof mw === "function") {
                this.#middleware.push(mw);
            }
        }
        return this;
    }
    async #stopStreams() {
        try {
            await this.#conversationsStream?.end();
        }
        finally {
            this.#conversationsStream = undefined;
        }
        try {
            await this.#messageStream?.end();
        }
        finally {
            this.#messageStream = undefined;
        }
    }
    /**
     * Closes all existing streams and restarts the streaming system.
     */
    async #handleStreamError(error) {
        await this.#stopStreams();
        const recovered = await this.#runErrorChain(error, {
            client: this.#client,
        });
        if (recovered) {
            this.#isLocked = false;
            queueMicrotask(() => this.start());
        }
    }
    async start(options) {
        if (this.#isLocked || this.#conversationsStream || this.#messageStream)
            return;
        this.#isLocked = true;
        try {
            this.#conversationsStream = await this.#client.conversations.stream({
                ...options,
                onValue: async (conversation) => {
                    try {
                        if (!conversation) {
                            return;
                        }
                        this.emit("conversation", new ConversationContext({
                            conversation,
                            client: this.#client,
                        }));
                        if (conversation instanceof Group) {
                            this.emit("group", new ConversationContext({
                                conversation,
                                client: this.#client,
                            }));
                        }
                        else if (conversation instanceof Dm) {
                            this.emit("dm", new ConversationContext({
                                conversation,
                                client: this.#client,
                            }));
                        }
                    }
                    catch (error) {
                        const recovered = await this.#runErrorChain(new AgentError(1001, "Emitted value from conversation stream caused an error.", error), new ClientContext({ client: this.#client }));
                        if (!recovered)
                            await this.stop();
                    }
                },
                onError: async (error) => {
                    const recovered = await this.#runErrorChain(new AgentStreamingError(1002, "Error occured during conversation streaming.", error), new ClientContext({ client: this.#client }));
                    if (!recovered)
                        await this.stop();
                },
            });
            this.#messageStream = await this.#client.conversations.streamAllMessages({
                ...options,
                onValue: async (message) => {
                    // this case should not happen,
                    // but we must handle it for proper types
                    if (!(message instanceof DecodedMessage)) {
                        return;
                    }
                    try {
                        switch (true) {
                            case isActions(message):
                                await this.#processMessage(message, "actions");
                                break;
                            case isAttachment(message):
                                await this.#processMessage(message, "inline-attachment");
                                break;
                            case isIntent(message):
                                await this.#processMessage(message, "intent");
                                break;
                            case isGroupUpdated(message):
                                await this.#processMessage(message, "group-update");
                                break;
                            case isLeaveRequest(message):
                                await this.#processMessage(message, "leave-request");
                                break;
                            case isMultiRemoteAttachment(message):
                                await this.#processMessage(message, "multi-attachment");
                                break;
                            case isRemoteAttachment(message):
                                await this.#processMessage(message, "attachment");
                                break;
                            case isReaction(message):
                                await this.#processMessage(message, "reaction");
                                break;
                            case isReadReceipt(message):
                                await this.#processMessage(message, "read-receipt");
                                break;
                            case isReply(message):
                                await this.#processMessage(message, "reply");
                                break;
                            case isTransactionReference(message):
                                await this.#processMessage(message, "transaction-reference");
                                break;
                            case isWalletSendCalls(message):
                                await this.#processMessage(message, "wallet-send-calls");
                                break;
                            case isMarkdown(message):
                                await this.#processMessage(message, "markdown");
                                break;
                            case isText(message):
                                await this.#processMessage(message, "text");
                                break;
                            default:
                                await this.#processMessage(message);
                                break;
                        }
                    }
                    catch (error) {
                        const recovered = await this.#runErrorChain(error, {
                            client: this.#client,
                        });
                        if (!recovered) {
                            await this.stop();
                        }
                        this.#isLocked = false;
                    }
                },
                onError: async (error) => {
                    const recovered = await this.#runErrorChain(new AgentStreamingError(1004, "Error occured during message streaming.", error), new ClientContext({ client: this.#client }));
                    if (!recovered)
                        await this.stop();
                },
            });
            this.emit("start", new ClientContext({ client: this.#client }));
            this.#isLocked = false;
        }
        catch (error) {
            await this.#handleStreamError(error);
        }
    }
    async #processMessage(message, topic = "unknownMessage") {
        // Skip messages with undefined content (failed to decode)
        if (!filter.hasContent(message)) {
            return;
        }
        // Skip messages from agent itself
        if (filter.fromSelf(message, this.#client)) {
            return;
        }
        const conversation = await this.#client.conversations.getConversationById(message.conversationId);
        if (!conversation) {
            throw new AgentError(1003, `Failed to process message ID "${message.id}" for conversation ID "${message.conversationId}" because the conversation could not be found.`);
        }
        const context = new MessageContext({
            message,
            conversation,
            client: this.#client,
        });
        await this.#runMiddlewareChain(context, topic);
    }
    async #runMiddlewareChain(context, topic = "unknownMessage") {
        const finalEmit = async () => {
            try {
                this.emit(topic, context);
                this.emit("message", context);
            }
            catch (error) {
                await this.#runErrorChain(error, context);
            }
        };
        const chain = this.#middleware.reduceRight((next, mw) => {
            return async () => {
                try {
                    await mw(context, next);
                }
                catch (error) {
                    const resume = await this.#runErrorChain(error, context);
                    if (resume) {
                        await next();
                    }
                    // Chain is not resuming, error is being swallowed
                }
            };
        }, finalEmit);
        await chain();
    }
    async #runErrorHandler(handler, context, error) {
        let settled = false;
        let flow = { kind: "stopped" };
        const next = (nextErr) => {
            if (settled)
                return;
            settled = true;
            flow =
                nextErr === undefined
                    ? { kind: "handled" }
                    : { kind: "continue", error: nextErr };
        };
        try {
            await handler(error, context, next);
            return flow;
        }
        catch (thrown) {
            if (settled) {
                return flow;
            }
            return { kind: "continue", error: thrown };
        }
    }
    async #runErrorChain(error, context) {
        const chain = [...this.#errorMiddleware, this.#defaultErrorHandler];
        let currentError = error;
        for (let i = 0; i < chain.length; i++) {
            const handler = chain[i];
            if (!handler)
                continue;
            const outcome = await this.#runErrorHandler(handler, context, currentError);
            switch (outcome.kind) {
                case "handled":
                    // Error was handled. Main middleware can continue.
                    return true;
                case "stopped":
                    // Error cannot be handled. Main middleware won't continue.
                    return false;
                case "continue":
                    // Error is passed to the next handler
                    currentError = outcome.error;
            }
        }
        // Reached end of chain without recovery
        return false;
    }
    get client() {
        return this.#client;
    }
    get errors() {
        return this.#errors;
    }
    async stop() {
        this.#isLocked = true;
        await this.#stopStreams();
        this.emit("stop", new ClientContext({ client: this.#client }));
        this.#isLocked = false;
    }
    createDmWithAddress(address, options) {
        return this.#client.conversations.createDmWithIdentifier({
            identifier: address,
            identifierKind: 0 /* IdentifierKind.Ethereum */,
        }, options);
    }
    createGroupWithAddresses(addresses, options) {
        const identifiers = addresses.map((address) => {
            return {
                identifier: address,
                identifierKind: 0 /* IdentifierKind.Ethereum */,
            };
        });
        return this.#client.conversations.createGroupWithIdentifiers(identifiers, options);
    }
    addMembersWithAddresses(group, addresses) {
        const identifiers = addresses.map((address) => {
            return {
                identifier: address,
                identifierKind: 0 /* IdentifierKind.Ethereum */,
            };
        });
        return group.addMembersByIdentifiers(identifiers);
    }
    async getConversationContext(conversationId) {
        const conversation = await this.client.conversations.getConversationById(conversationId);
        if (conversation) {
            const context = new ConversationContext({
                conversation,
                client: this.#client,
            });
            return context;
        }
    }
    get address() {
        return this.#client.accountIdentifier?.identifier;
    }
}

class CommandRouter {
    #commandMap = new Map();
    #defaultHandler = null;
    constructor(config = {}) {
        if (config.helpCommand) {
            this.#registerHelpCommand(config.helpCommand);
        }
    }
    #registerHelpCommand(command) {
        const helpHandler = async (ctx) => {
            const lines = [];
            const sortedCommands = [...this.#commandMap.entries()].sort((a, b) => a[0].localeCompare(b[0]));
            sortedCommands.forEach(([cmd, entry], index) => {
                if (entry.description) {
                    lines.push(`${index + 1}. \`${cmd}\` - ${entry.description}`);
                }
                else {
                    lines.push(`${index + 1}. \`${cmd}\``);
                }
            });
            await ctx.sendMarkdownReply(lines.join("\n"));
        };
        this.command(command, "Show available commands", helpHandler);
    }
    get commandList() {
        return Array.from(this.#commandMap.keys());
    }
    command(command, handlerOrDescription, handler) {
        if (!command.startsWith("/")) {
            throw new Error('Command must start with "/"');
        }
        let resolvedHandler;
        let description;
        if (typeof handlerOrDescription === "function") {
            resolvedHandler = handlerOrDescription;
        }
        else {
            description = handlerOrDescription;
            if (!handler) {
                throw new Error("Handler implementation is required when description is provided.");
            }
            resolvedHandler = handler;
        }
        this.#commandMap.set(command.toLowerCase(), {
            handler: resolvedHandler,
            description,
        });
        return this;
    }
    default(handler) {
        this.#defaultHandler = handler;
        return this;
    }
    async handle(ctx) {
        const messageText = ctx.message.content;
        const parts = messageText.split(" ");
        const command = parts[0]?.toLowerCase();
        if (!command) {
            return false;
        }
        // Check if this is a command message
        if (command.startsWith("/")) {
            const entry = this.#commandMap.get(command);
            if (entry) {
                // Create a new context with modified content (everything after the command)
                const argsText = parts.slice(1).join(" ");
                ctx.message.content = argsText;
                await entry.handler(ctx);
                return true;
            }
        }
        // If no command matched and there's a default handler, use it
        if (this.#defaultHandler) {
            await this.#defaultHandler(ctx);
            return true;
        }
        return false;
    }
    middleware() {
        return async (ctx, next) => {
            if (ctx.isText()) {
                const handled = await this.handle(ctx);
                if (!handled) {
                    await next();
                }
            }
            else {
                await next();
            }
        };
    }
}

/**
 * Middleware that measures message processing time and logs periodic
 * CPU / memory health reports.
 *
 * Register it as the first middleware so the timer wraps all downstream
 * middleware and handlers, giving you the total processing time independent
 * of other logic.
 */
class PerformanceMonitor {
    #interval;
    #isShutdown;
    #lastCpuUsage;
    #lastCpuTime;
    #criticalThresholdInterval;
    #onCriticalResponse;
    #onHealthReport;
    #onResponse;
    #onShutdown;
    #eventLoopHistogram;
    constructor(config = {}) {
        const { healthReportInterval = 60_000, criticalThresholdInterval = 10_000, } = config;
        this.#criticalThresholdInterval = criticalThresholdInterval;
        const defaultCriticalResponseHandler = (durationMs) => {
            console.warn(`[PerformanceMonitor] Critical: Response time exceeded ${criticalThresholdInterval / 1000}s (${durationMs.toFixed(0)}ms)`);
        };
        const defaultHealthReportHandler = (report) => {
            console.log(`[${new Date().toISOString()}] CPU: ${report.cpuPercent.toFixed(1)}% | Event Loop: ${report.eventLoopDelayMs.toFixed(1)}ms | Heap: ${report.heapPercent.toFixed(1)}% (${report.heapMB.toFixed(0)}MB/${report.totalMB.toFixed(1)}MB)`);
        };
        const defaultShutdownHandler = () => {
            console.log("[PerformanceMonitor] Monitoring shut down");
        };
        this.#onCriticalResponse =
            config.onCriticalResponse ?? defaultCriticalResponseHandler;
        this.#onHealthReport = config.onHealthReport ?? defaultHealthReportHandler;
        this.#onResponse = config.onResponse;
        this.#onShutdown = config.onShutdown ?? defaultShutdownHandler;
        this.#isShutdown = false;
        this.#eventLoopHistogram = monitorEventLoopDelay();
        this.#eventLoopHistogram.enable();
        this.#lastCpuUsage = process.cpuUsage();
        this.#lastCpuTime = Date.now();
        if (healthReportInterval > 0) {
            this.#logHealthReport();
            this.#interval = setInterval(() => {
                this.#logHealthReport();
            }, healthReportInterval).unref();
        }
    }
    #getCpuPercent() {
        const currentCpuUsage = process.cpuUsage(this.#lastCpuUsage);
        const elapsedMs = Date.now() - this.#lastCpuTime;
        const totalCpuMicros = currentCpuUsage.user + currentCpuUsage.system;
        const elapsedMicros = elapsedMs * 1000;
        const percent = elapsedMicros > 0 ? (totalCpuMicros / elapsedMicros) * 100 : 0;
        this.#lastCpuUsage = process.cpuUsage();
        this.#lastCpuTime = Date.now();
        return percent;
    }
    #getMemory() {
        const memoryUsage = process.memoryUsage();
        const heapStats = v8.getHeapStatistics();
        const heapUsedMB = memoryUsage.heapUsed / 1024 / 1024;
        const heapLimitMB = heapStats.heap_size_limit / 1024 / 1024;
        return {
            heapMB: heapUsedMB,
            heapPercent: (heapUsedMB / heapLimitMB) * 100,
            heapLimitMB,
            totalMB: memoryUsage.rss / 1024 / 1024,
        };
    }
    #logHealthReport() {
        const cpuPercent = this.#getCpuPercent();
        const mem = this.#getMemory();
        const mean = this.#eventLoopHistogram.mean;
        const eventLoopDelayMs = Number.isNaN(mean) ? 0 : mean / 1e6;
        this.#eventLoopHistogram.reset();
        this.#onHealthReport({
            cpuPercent,
            eventLoopDelayMs,
            ...mem,
        });
    }
    shutdown() {
        if (this.#isShutdown) {
            return;
        }
        this.#isShutdown = true;
        clearInterval(this.#interval);
        this.#interval = undefined;
        this.#eventLoopHistogram.disable();
        this.#onShutdown();
    }
    middleware() {
        return async (_, next) => {
            const start = performance.now();
            try {
                await next();
            }
            finally {
                const duration = performance.now() - start;
                this.#onResponse?.(duration);
                if (duration > this.#criticalThresholdInterval) {
                    this.#onCriticalResponse(duration);
                }
            }
        };
    }
}

class LimitedMap {
    #map = new Map();
    #limit;
    constructor(limit) {
        this.#limit = limit;
    }
    set(key, value) {
        if (this.#map.size >= this.#limit) {
            const it = this.#map.keys().next();
            if (!it.done) {
                this.#map.delete(it.value);
            }
        }
        this.#map.set(key, value);
    }
    get(key) {
        return this.#map.get(key);
    }
}

const cache = new LimitedMap(1000);
const fetchFromWeb3Bio = async (name, apiKey) => {
    const endpoint = `https://api.web3.bio/ns/${escape(name)}`;
    const headers = {
        "Content-Type": "application/json",
    };
    if (apiKey) {
        headers["X-API-KEY"] = `Bearer ${apiKey}`;
    }
    const response = await fetch(endpoint, {
        method: "GET",
        headers,
    });
    if (!response.ok) {
        throw new AgentError(2000, `Could not resolve address for name "${name}": ${response.statusText} (${response.status})`);
    }
    return response.json();
};
const resolveName = async (name, apiKey) => {
    if (isAddress(name)) {
        return name;
    }
    const cachedAddress = cache.get(name);
    if (cachedAddress !== undefined) {
        return cachedAddress;
    }
    const response = await fetchFromWeb3Bio(name, apiKey);
    if (response.length === 0) {
        return null;
    }
    const address = response[0]?.address;
    if (!address) {
        return null;
    }
    cache.set(name, address);
    return address;
};
const createNameResolver = (apiKey) => {
    return (name) => resolveName(name, apiKey);
};

/**
 * Minimal ERC-20 ABI containing transfer, balanceOf, and decimals functions.
 * Can be used with viem's encodeFunctionData for custom ERC-20 interactions.
 *
 * @see https://eips.ethereum.org/EIPS/eip-20#methods
 */
const erc20Abi = [
    {
        type: "function",
        name: "transfer",
        inputs: [
            { name: "to", type: "address" },
            { name: "amount", type: "uint256" },
        ],
        outputs: [{ name: "", type: "bool" }],
        stateMutability: "nonpayable",
    },
    {
        type: "function",
        name: "balanceOf",
        inputs: [{ name: "account", type: "address" }],
        outputs: [{ name: "", type: "uint256" }],
        stateMutability: "view",
    },
    {
        type: "function",
        name: "decimals",
        inputs: [],
        outputs: [{ name: "", type: "uint8" }],
        stateMutability: "view",
    },
];
/**
 * Creates a WalletSendCalls payload for an ERC-20 token transfer.
 *
 * @param options - The transfer options
 * @returns A WalletSendCalls object ready to send
 */
function createERC20TransferCalls(options) {
    const { chain, tokenAddress, from, to, amount, description } = options;
    const data = encodeFunctionData({
        abi: erc20Abi,
        functionName: "transfer",
        args: [to, amount],
    });
    return {
        version: "1.0",
        chainId: toHex(chain.id),
        from,
        calls: [
            {
                to: tokenAddress,
                data,
                value: "0x0",
                metadata: {
                    description,
                    transactionType: "transfer",
                },
            },
        ],
    };
}
/**
 * Creates a WalletSendCalls payload for a native token transfer (ETH, MATIC, etc.).
 *
 * @param options - The transfer options
 * @returns A WalletSendCalls object ready to send
 */
function createNativeTransferCalls(options) {
    const { chain, from, to, amount, description } = options;
    return {
        version: "1.0",
        chainId: toHex(chain.id),
        from,
        calls: [
            {
                to,
                value: toHex(amount),
                metadata: {
                    description,
                    transactionType: "transfer",
                },
            },
        ],
    };
}
/**
 * Reads the ERC-20 token balance for a given address from the blockchain.
 *
 * @param options - The query options including chain, token address, and wallet address
 * @returns The token balance in base units as a bigint
 */
async function getERC20Balance(options) {
    const { chain, tokenAddress, address, transport: customTransport } = options;
    const client = createPublicClient({
        chain,
        transport: customTransport ?? http(),
    });
    return client.readContract({
        address: tokenAddress,
        abi: erc20Abi,
        functionName: "balanceOf",
        args: [address],
    });
}
/**
 * Reads the number of decimals for an ERC-20 token from the blockchain.
 *
 * @param options - The query options including chain and token address
 * @returns The number of decimals (typically 6 or 18)
 */
async function getERC20Decimals(options) {
    const { chain, tokenAddress, transport: customTransport } = options;
    const client = createPublicClient({
        chain,
        transport: customTransport ?? http(),
    });
    return client.readContract({
        address: tokenAddress,
        abi: erc20Abi,
        functionName: "decimals",
    });
}

export { Agent, AgentError, AgentStreamingError, ClientContext, CommandRouter, ConversationContext, MessageContext, PerformanceMonitor, createERC20TransferCalls, createIdentifier, createNameResolver, createNativeTransferCalls, createRemoteAttachment, createRemoteAttachmentFromFile, createSigner, createUser, downloadRemoteAttachment, erc20Abi, f, filter, getERC20Balance, getERC20Decimals, getInstallationInfo, getTestUrl, getValidLogLevels, logDetails, parseLogLevel };
