// @azeth/common@0.2.0/dist/constants/addresses.js
// Contract addresses — auto-generated by scripts/sync-deployments.ts
// Do not edit manually. Run: pnpm sync:deployments
export const AZETH_CONTRACTS = {
    baseSepolia: {
        factory: '0x8B8cc09732cDF3e18EF441F70f88DfC0D11EeD3f',
        guardianModule: '0xBC11De1fbA08D5d9294d1540A322dc1c8279CC3B',
        trustRegistryModule: '0x106c2E84c00fdD32383cA315E43976aaA7930C03',
        paymentAgreementModule: '0x566007ae53028b85699400CD437E011F5975cde9',
        reputationModule: '0x0dC17903834b1cd704d3acb1f0c6f8A6529be2d5',
        priceOracle: '0xFAeBC147a10F73669dD47a5CFC785f04aa47E696',
        accountImplementation: '0x229dc447c874D19927da2B4B76dE97D8b3964Fcc',
    },
    ethereumSepolia: {
        factory: '0x8B8cc09732cDF3e18EF441F70f88DfC0D11EeD3f',
        guardianModule: '0xBC11De1fbA08D5d9294d1540A322dc1c8279CC3B',
        trustRegistryModule: '0x106c2E84c00fdD32383cA315E43976aaA7930C03',
        paymentAgreementModule: '0x566007ae53028b85699400CD437E011F5975cde9',
        reputationModule: '0x0dC17903834b1cd704d3acb1f0c6f8A6529be2d5',
        priceOracle: '0xFAeBC147a10F73669dD47a5CFC785f04aa47E696',
        accountImplementation: '0x229dc447c874D19927da2B4B76dE97D8b3964Fcc',
    },
    base: {
        factory: '',
        guardianModule: '',
        trustRegistryModule: '',
        paymentAgreementModule: '',
        reputationModule: '',
        priceOracle: '',
        accountImplementation: '',
    },
    ethereum: {
        factory: '',
        guardianModule: '',
        trustRegistryModule: '',
        paymentAgreementModule: '',
        reputationModule: '',
        priceOracle: '',
        accountImplementation: '',
    },
};
// External contracts — ERC-8004 Identity Registry
export const ERC8004_REGISTRY = {
    base: '0x8004A169FB4a3325136EB29fA0ceB6D2e539a432',
    ethereum: '0x8004A169FB4a3325136EB29fA0ceB6D2e539a432',
    baseSepolia: '0x8004A818BFB912233c491871b3d84c89A494BD9e',
    ethereumSepolia: '0x8004A818BFB912233c491871b3d84c89A494BD9e',
};
// External contracts — ERC-8004 Reputation Registry
export const ERC8004_REPUTATION_REGISTRY = {
    base: '0x8004BAa17C55a88189AE136b182e5fdA19dE9b63',
    ethereum: '0x8004BAa17C55a88189AE136b182e5fdA19dE9b63',
    baseSepolia: '0x8004B663056A597Dffe9eCcC1965A193B7388713',
    ethereumSepolia: '0x8004B663056A597Dffe9eCcC1965A193B7388713',
};
export const ENTRYPOINT_V07 = '0x0000000071727De22E5E9d8BAf0edAc6f37da032';
