// @azeth/common@0.2.0/dist/constants/chains.js
export const SUPPORTED_CHAINS = {
    base: {
        id: 8453,
        name: 'Base',
        rpcDefault: 'https://mainnet.base.org',
        bundlerBaseUrl: 'https://api.pimlico.io/v2/8453/rpc',
    },
    baseSepolia: {
        id: 84532,
        name: 'Base Sepolia',
        rpcDefault: 'https://sepolia.base.org',
        bundlerBaseUrl: 'https://api.pimlico.io/v2/84532/rpc',
    },
    ethereumSepolia: {
        id: 11155111,
        name: 'Ethereum Sepolia',
        rpcDefault: 'https://rpc.sepolia.org',
        bundlerBaseUrl: 'https://api.pimlico.io/v2/11155111/rpc',
    },
    ethereum: {
        id: 1,
        name: 'Ethereum',
        rpcDefault: 'https://cloudflare-eth.com',
        bundlerBaseUrl: 'https://api.pimlico.io/v2/1/rpc',
    },
};
/** Build a full bundler URL from the chain's base URL and an API key.
 *  Returns undefined if either is missing.
 *
 *  @param chainName - Chain to look up (e.g. 'baseSepolia')
 *  @param apiKey - Pimlico API key (from PIMLICO_API_KEY env var)
 */
export function getBundlerUrl(chainName, apiKey) {
    const chain = SUPPORTED_CHAINS[chainName];
    if (!chain?.bundlerBaseUrl || !apiKey)
        return undefined;
    return `${chain.bundlerBaseUrl}?apikey=${apiKey}`;
}
/** Build a full paymaster URL from the chain's base URL and an API key.
 *  Returns undefined if either is missing.
 *
 *  Pimlico uses the same endpoint for both bundler and paymaster RPCs —
 *  the differentiation is at the RPC method level (`pm_sponsorUserOperation`
 *  vs `eth_sendUserOperation`), not the URL level.
 *
 *  @param chainName - Chain to look up (e.g. 'baseSepolia')
 *  @param apiKey - Pimlico API key (from PIMLICO_API_KEY env var)
 */
/** Build the Azeth server bundler proxy URL from a server URL.
 *
 *  Used as a fallback when no PIMLICO_API_KEY is available — the Azeth server
 *  proxies bundler requests using its own API key (testnet only).
 *
 *  @param serverUrl - Azeth server URL (e.g. 'https://api.azeth.ai')
 */
export function getServerBundlerUrl(serverUrl) {
    return `${serverUrl.replace(/\/$/, '')}/api/v1/bundler/rpc`;
}
export function getPaymasterUrl(chainName, apiKey) {
    const chain = SUPPORTED_CHAINS[chainName];
    if (!chain?.bundlerBaseUrl || !apiKey)
        return undefined;
    // Pimlico shares the same v2 endpoint for both bundler and paymaster
    return `${chain.bundlerBaseUrl}?apikey=${apiKey}`;
}
