// @xmtp/node-sdk@5.3.0/dist/index.js
import { generateInboxId as generateInboxId$1, getInboxIdForIdentifier as getInboxIdForIdentifier$1, contentTypeIntent, contentTypeActions, contentTypeWalletSendCalls, contentTypeLeaveRequest, contentTypeReadReceipt, contentTypeGroupUpdated, contentTypeTransactionReference, contentTypeMultiRemoteAttachment, contentTypeRemoteAttachment, contentTypeAttachment, contentTypeReaction, contentTypeReply, contentTypeMarkdown, contentTypeText, createClient as createClient$1, revokeInstallationsSignatureRequest, applySignatureRequest, inboxStateFromInboxIds, verifySignedWithPublicKey, isAddressAuthorized, isInstallationAuthorized } from '@xmtp/node-bindings';
export { ActionStyle, ConsentEntityType, ConsentState, ContentType, ConversationType, DeliveryStatus, GroupMembershipState, GroupMessageKind, GroupPermissionsOptions, IdentifierKind, ListConversationsOrderBy, LogLevel, MessageSortBy, MetadataField, PermissionLevel, PermissionPolicy, PermissionUpdateType, ReactionAction, ReactionSchema, SortDirection, contentTypeActions, contentTypeAttachment, contentTypeGroupUpdated, contentTypeIntent, contentTypeLeaveRequest, contentTypeMarkdown, contentTypeMultiRemoteAttachment, contentTypeReaction, contentTypeReadReceipt, contentTypeRemoteAttachment, contentTypeReply, contentTypeText, contentTypeTransactionReference, contentTypeWalletSendCalls, decryptAttachment, encodeActions, encodeAttachment, encodeIntent, encodeMarkdown, encodeMultiRemoteAttachment, encodeReaction, encodeReadReceipt, encodeRemoteAttachment, encodeText, encodeTransactionReference, encodeWalletSendCalls, encryptAttachment } from '@xmtp/node-bindings';
import { contentTypeToString } from '@xmtp/content-type-primitives';
import { isPromise } from 'node:util/types';
import { join } from 'node:path';
import process from 'node:process';

/**
 * Pre-configured URLs for the XMTP network based on the environment
 *
 * @constant
 * @property {string} local - The local URL for the XMTP network
 * @property {string} dev - The development URL for the XMTP network
 * @property {string} production - The production URL for the XMTP network
 */
const ApiUrls = {
    local: "http://localhost:5556",
    dev: "https://grpc.dev.xmtp.network:443",
    production: "https://grpc.production.xmtp.network:443",
};
/**
 * Pre-configured URLs for the XMTP history sync service based on the environment
 *
 * @constant
 * @property {string} local - The local URL for the XMTP history sync service
 * @property {string} dev - The development URL for the XMTP history sync service
 * @property {string} production - The production URL for the XMTP history sync service
 */
const HistorySyncUrls = {
    local: "http://localhost:5558",
    dev: "https://message-history.dev.ephemera.network",
    production: "https://message-history.production.ephemera.network",
};

class InboxReassignError extends Error {
    constructor() {
        super("Unable to create add account signature text, `allowInboxReassign` must be true");
    }
}
class AccountAlreadyAssociatedError extends Error {
    constructor(inboxId) {
        super(`Account already associated with inbox ${inboxId}`);
    }
}
class MissingContentTypeError extends Error {
    constructor() {
        super("Content type is required when sending encoded content");
    }
}
class SignerUnavailableError extends Error {
    constructor() {
        super("Signer unavailable, use Client.create to create a client with a signer");
    }
}
class ClientNotInitializedError extends Error {
    constructor() {
        super("Client not initialized, use Client.create or Client.build to create a client");
    }
}
class StreamFailedError extends Error {
    constructor(retryAttempts) {
        const times = `time${retryAttempts !== 1 ? "s" : ""}`;
        super(`Stream failed, retried ${retryAttempts} ${times}`);
    }
}
class StreamInvalidRetryAttemptsError extends Error {
    constructor() {
        super("Stream retry attempts must be greater than 0");
    }
}

const generateInboxId = (identifier, nonce) => {
    return generateInboxId$1(identifier, nonce);
};
const getInboxIdForIdentifier = async (identifier, env = "dev", gatewayHost) => {
    const host = ApiUrls[env];
    const isSecure = host.startsWith("https");
    return getInboxIdForIdentifier$1(host, gatewayHost, isSecure, identifier);
};

function isHexString(value) {
    return typeof value === "string" && /^0x(?:[0-9a-fA-F]{2})+$/.test(value);
}
function validHex(value) {
    if (!isHexString(value)) {
        throw new TypeError(`Value is not a hexadecimal string.`);
    }
    return value;
}

const isReaction = (m) => m.contentType.authorityId === "xmtp.org" &&
    m.contentType.typeId === "reaction";
const isReply = (m) => m.contentType.authorityId === "xmtp.org" && m.contentType.typeId === "reply";
const isTextReply = (m) => isReply(m) && typeof m.content?.content === "string";
const isText = (m) => m.contentType.authorityId === "xmtp.org" && m.contentType.typeId === "text";
const isRemoteAttachment = (m) => m.contentType.authorityId === "xmtp.org" &&
    m.contentType.typeId === "remoteStaticAttachment";
const isAttachment = (m) => m.contentType.authorityId === "xmtp.org" &&
    m.contentType.typeId === "attachment";
const isMultiRemoteAttachment = (m) => m.contentType.authorityId === "xmtp.org" &&
    m.contentType.typeId === "multiRemoteStaticAttachment";
const isTransactionReference = (m) => m.contentType.authorityId === "xmtp.org" &&
    m.contentType.typeId === "transactionReference";
const isGroupUpdated = (m) => m.contentType.authorityId === "xmtp.org" &&
    m.contentType.typeId === "group_updated";
const isReadReceipt = (m) => m.contentType.authorityId === "xmtp.org" &&
    m.contentType.typeId === "readReceipt";
const isLeaveRequest = (m) => m.contentType.authorityId === "xmtp.org" &&
    m.contentType.typeId === "leave_request";
const isWalletSendCalls = (m) => m.contentType.authorityId === "xmtp.org" &&
    m.contentType.typeId === "walletSendCalls";
const isIntent = (m) => m.contentType.authorityId === "coinbase.com" &&
    m.contentType.typeId === "intent";
const isActions = (m) => m.contentType.authorityId === "coinbase.com" &&
    m.contentType.typeId === "actions";
const isMarkdown = (m) => m.contentType.authorityId === "xmtp.org" &&
    m.contentType.typeId === "markdown";

class CodecRegistry {
    #codecs;
    constructor(codecs) {
        this.#codecs = new Map(codecs.map((codec) => [contentTypeToString(codec.contentType), codec]));
    }
    /**
     * Gets the codec for a given content type
     *
     * @param contentType - The content type to get the codec for
     * @returns The codec, if found
     */
    getCodec(contentType) {
        return this.#codecs.get(contentTypeToString(contentType));
    }
}

function nsToDate(ns) {
    return new Date(Number(ns / 1000000n));
}

const getContentFromDecodedMessageContent = (content) => {
    switch (content.type) {
        case "Text" /* DecodedMessageContentType.Text */: {
            return content.text;
        }
        case "Markdown" /* DecodedMessageContentType.Markdown */: {
            return content.markdown;
        }
        case "Reply" /* DecodedMessageContentType.Reply */: {
            return content.reply;
        }
        case "Reaction" /* DecodedMessageContentType.Reaction */: {
            return content.reaction;
        }
        case "Attachment" /* DecodedMessageContentType.Attachment */: {
            return content.attachment;
        }
        case "RemoteAttachment" /* DecodedMessageContentType.RemoteAttachment */: {
            return content.remoteAttachment;
        }
        case "MultiRemoteAttachment" /* DecodedMessageContentType.MultiRemoteAttachment */: {
            return content.multiRemoteAttachment;
        }
        case "TransactionReference" /* DecodedMessageContentType.TransactionReference */: {
            return content.transactionReference;
        }
        case "GroupUpdated" /* DecodedMessageContentType.GroupUpdated */: {
            return content.groupUpdated;
        }
        case "ReadReceipt" /* DecodedMessageContentType.ReadReceipt */: {
            return content.readReceipt;
        }
        case "LeaveRequest" /* DecodedMessageContentType.LeaveRequest */: {
            return content.leaveRequest;
        }
        case "WalletSendCalls" /* DecodedMessageContentType.WalletSendCalls */: {
            return content.walletSendCalls;
        }
        case "Actions" /* DecodedMessageContentType.Actions */: {
            return content.actions;
        }
        case "Intent" /* DecodedMessageContentType.Intent */: {
            return content.intent;
        }
        case "DeletedMessage" /* DecodedMessageContentType.DeletedMessage */: {
            return content.deletedMessage;
        }
        case "Custom" /* DecodedMessageContentType.Custom */: {
            return content.custom;
        }
        default:
            content.type;
            return null;
    }
};
const getContentTypeFromDecodedMessageContent = (content) => {
    switch (content.type) {
        case "Text" /* DecodedMessageContentType.Text */: {
            return contentTypeText();
        }
        case "Markdown" /* DecodedMessageContentType.Markdown */: {
            return contentTypeMarkdown();
        }
        case "Reply" /* DecodedMessageContentType.Reply */: {
            return contentTypeReply();
        }
        case "Reaction" /* DecodedMessageContentType.Reaction */: {
            return contentTypeReaction();
        }
        case "Attachment" /* DecodedMessageContentType.Attachment */: {
            return contentTypeAttachment();
        }
        case "RemoteAttachment" /* DecodedMessageContentType.RemoteAttachment */: {
            return contentTypeRemoteAttachment();
        }
        case "MultiRemoteAttachment" /* DecodedMessageContentType.MultiRemoteAttachment */: {
            return contentTypeMultiRemoteAttachment();
        }
        case "TransactionReference" /* DecodedMessageContentType.TransactionReference */: {
            return contentTypeTransactionReference();
        }
        case "GroupUpdated" /* DecodedMessageContentType.GroupUpdated */: {
            return contentTypeGroupUpdated();
        }
        case "ReadReceipt" /* DecodedMessageContentType.ReadReceipt */: {
            return contentTypeReadReceipt();
        }
        case "LeaveRequest" /* DecodedMessageContentType.LeaveRequest */: {
            return contentTypeLeaveRequest();
        }
        case "WalletSendCalls" /* DecodedMessageContentType.WalletSendCalls */: {
            return contentTypeWalletSendCalls();
        }
        case "Actions" /* DecodedMessageContentType.Actions */: {
            return contentTypeActions();
        }
        case "Intent" /* DecodedMessageContentType.Intent */: {
            return contentTypeIntent();
        }
        case "DeletedMessage" /* DecodedMessageContentType.DeletedMessage */: {
            return undefined;
        }
        case "Custom" /* DecodedMessageContentType.Custom */: {
            return content.custom?.type;
        }
        default:
            content.type;
            return undefined;
    }
};
/**
 * Represents a decoded XMTP message
 *
 * @class
 * @property {unknown} content - The decoded content of the message
 * @property {ContentTypeId} contentType - The content type of the message content
 * @property {string} conversationId - Unique identifier for the conversation
 * @property {MessageDeliveryStatus} deliveryStatus - Current delivery status of the message ("unpublished" | "published" | "failed")
 * @property {bigint} expiresAtNs - Timestamp when the message will expire (in nanoseconds)
 * @property {Date} expiresAt - Timestamp when the message will expire
 * @property {string} [fallback] - Optional fallback text for the message
 * @property {string} id - Unique identifier for the message
 * @property {MessageKind} kind - Type of message ("application" | "membership_change")
 * @property {number} numReplies - Number of replies to the message
 * @property {DecodedMessage<Reaction>[]} reactions - Reactions to the message
 * @property {string} senderInboxId - Identifier for the sender's inbox
 * @property {Date} sentAt - Timestamp when the message was sent
 * @property {bigint} sentAtNs - Timestamp when the message was sent (in nanoseconds)
 */
class DecodedMessage {
    content;
    contentType;
    conversationId;
    deliveryStatus;
    expiresAtNs;
    expiresAt;
    fallback;
    id;
    kind;
    numReplies;
    reactions;
    senderInboxId;
    sentAt;
    sentAtNs;
    constructor(codecRegistry, message) {
        this.id = message.id;
        this.expiresAtNs = message.expiresAtNs ?? undefined;
        this.expiresAt = message.expiresAtNs
            ? nsToDate(message.expiresAtNs)
            : undefined;
        this.sentAtNs = message.sentAtNs;
        this.sentAt = nsToDate(message.sentAtNs);
        this.conversationId = message.conversationId;
        this.senderInboxId = message.senderInboxId;
        this.contentType = message.contentType;
        this.fallback = message.fallback ?? undefined;
        this.kind = message.kind;
        this.deliveryStatus = message.deliveryStatus;
        this.numReplies = message.numReplies;
        this.reactions = message.reactions.map((reaction) => new DecodedMessage(codecRegistry, reaction));
        this.content =
            getContentFromDecodedMessageContent(message.content) ??
                undefined;
        switch (message.content.type) {
            case "Reply" /* DecodedMessageContentType.Reply */: {
                const reply = message.content.reply;
                let replyContent = getContentFromDecodedMessageContent(reply.content);
                if (reply.content.type === "Custom" /* DecodedMessageContentType.Custom */) {
                    const codec = codecRegistry.getCodec(reply.content.custom?.type);
                    if (codec) {
                        try {
                            replyContent = codec.decode(replyContent);
                        }
                        catch (error) {
                            if (error instanceof Error) {
                                console.warn(`Error decoding custom content: ${error.message}`);
                            }
                            else {
                                console.warn(`Error decoding custom content`);
                            }
                        }
                    }
                }
                this.content = {
                    referenceId: reply.referenceId,
                    content: replyContent,
                    contentType: getContentTypeFromDecodedMessageContent(reply.content),
                    inReplyTo: reply.inReplyTo
                        ? new DecodedMessage(codecRegistry, reply.inReplyTo)
                        : null,
                };
                break;
            }
            case "Custom" /* DecodedMessageContentType.Custom */: {
                const customContent = message.content.custom;
                if (customContent !== null) {
                    const codec = codecRegistry.getCodec(this.contentType);
                    if (codec) {
                        try {
                            this.content = codec.decode(customContent);
                        }
                        catch (error) {
                            if (error instanceof Error) {
                                console.warn(`Error decoding custom content: ${error.message}`);
                            }
                            else {
                                console.warn(`Error decoding custom content`);
                            }
                            this.content = undefined;
                        }
                    }
                    else {
                        console.warn(`No codec found for content type "${contentTypeToString(this.contentType)}"`);
                        this.content = undefined;
                    }
                }
                break;
            }
        }
    }
}

/**
 * AsyncStream provides an async iterable interface for streaming data.
 *
 * This class implements a producer-consumer pattern where:
 * - Producers can push values using the `push()` method
 * - Consumers can iterate over values asynchronously using `for await` loops or `next()`
 * - Values are queued internally when no consumers are waiting
 * - Consumers are resolved immediately when values are available
 * - The stream can be terminated using `done()`, `return()`, or `end()`
 *
 * @example
 * ```typescript
 * const stream = new AsyncStream<string>();
 *
 * stream.push("hello");
 * stream.push("world");
 *
 * for await (const value of stream) {
 *   console.log(value); // "hello", "world"
 * }
 * ```
 */
class AsyncStream {
    isDone = false;
    #pendingResolves = [];
    #queue;
    onDone;
    onReturn;
    constructor() {
        this.#queue = [];
        this.isDone = false;
    }
    flush() {
        while (this.#pendingResolves.length > 0) {
            const nextResolve = this.#pendingResolves.shift();
            if (nextResolve) {
                nextResolve({ done: true, value: undefined });
            }
        }
    }
    done() {
        this.flush();
        this.#queue = [];
        this.#pendingResolves = [];
        this.isDone = true;
        this.onDone?.();
    }
    push = (value) => {
        if (this.isDone) {
            return;
        }
        const nextResolve = this.#pendingResolves.shift();
        if (nextResolve) {
            nextResolve({
                done: false,
                value,
            });
        }
        else {
            this.#queue.push(value);
        }
    };
    next = () => {
        if (this.isDone) {
            return Promise.resolve({ done: true, value: undefined });
        }
        if (this.#queue.length > 0) {
            return Promise.resolve({
                done: false,
                value: this.#queue.shift(),
            });
        }
        return new Promise((resolve) => {
            this.#pendingResolves.push(resolve);
        });
    };
    return = () => {
        this.onReturn?.();
        this.done();
        return Promise.resolve({
            done: true,
            value: undefined,
        });
    };
    end = () => this.return();
    [Symbol.asyncIterator]() {
        return this;
    }
}
const usableProperties = [
    "end",
    "isDone",
    "next",
    "return",
    Symbol.asyncIterator,
];
const isUsableProperty = (prop) => {
    return usableProperties.includes(prop);
};
/**
 * Creates a read-only proxy for AsyncStream instances that restricts access to consumer-only methods.
 *
 * This proxy only exposes the following properties and methods:
 * - `next()`: Get the next value from the stream
 * - `end()`: Terminate the stream and stop iteration
 * - `return()`: Same as end(), terminates the stream
 * - `isDone`: Boolean indicating if the stream has been terminated
 * - `Symbol.asyncIterator`: Enables `for await` loop iteration
 *
 * Producer methods like `push()`, `done()`, and `flush()` are hidden to prevent
 * consumers from accidentally modifying the stream state.
 *
 * @param stream - The AsyncStream instance to create a proxy for
 * @returns A read-only proxy that implements AsyncStreamProxy<T>
 *
 * @example
 * ```typescript
 * const stream = new AsyncStream<string>();
 * const proxy = createAsyncStreamProxy(stream);
 *
 * stream.push("hello");
 * stream.push("world");
 *
 * for await (const value of proxy) {
 *   console.log(value); // "hello", "world"
 * }
 * ```
 */
function createAsyncStreamProxy(stream) {
    return new Proxy(stream, {
        get(target, prop, receiver) {
            if (isUsableProperty(prop)) {
                return Reflect.get(target, prop, receiver);
            }
        },
        set() {
            return true;
        },
        has(_target, prop) {
            return isUsableProperty(prop);
        },
        ownKeys() {
            return usableProperties;
        },
        getOwnPropertyDescriptor(target, prop) {
            if (isUsableProperty(prop)) {
                return {
                    enumerable: true,
                    configurable: true,
                    value: Reflect.get(target, prop),
                };
            }
            return undefined;
        },
    });
}

const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
const DEFAULT_RETRY_DELAY = 60_000; // milliseconds
const DEFAULT_RETRY_ATTEMPTS = 10;
/**
 * Creates a stream from a stream function
 *
 * If the stream fails, an attempt will be made to restart it.
 *
 * This function is not intended to be used directly.
 *
 * @param streamFunction - The stream function to create a stream from
 * @param streamValueMutator - An optional function to mutate the value emitted from the stream
 * @param options - The options for the stream
 * @param args - Additional arguments to pass to the stream function
 * @returns An async iterable stream proxy
 * @throws {StreamInvalidRetryAttemptsError} if the retryAttempts option is less than 0 and retryOnFail is true
 * @throws {StreamFailedError} if the stream fails and can't be restarted
 */
const createStream = async (streamFunction, streamValueMutator, options) => {
    const { onError, onFail, onRestart, onRetry, onValue, retryAttempts = DEFAULT_RETRY_ATTEMPTS, retryDelay = DEFAULT_RETRY_DELAY, retryOnFail = true, } = options ?? {};
    // retry attempts must be greater than 0
    if (retryOnFail && retryAttempts < 0) {
        throw new StreamInvalidRetryAttemptsError();
    }
    const asyncStream = new AsyncStream();
    const streamCallback = (error, value) => {
        // if a stream error occurs, call the onError callback
        if (error) {
            onError?.(error);
            return;
        }
        // ensure the value is not undefined
        if (value !== undefined) {
            try {
                // if a streamValueMutator is provided, mutate the value
                if (streamValueMutator) {
                    const mutatedValue = streamValueMutator(value);
                    if (isPromise(mutatedValue)) {
                        void mutatedValue
                            .then((mutatedValue) => {
                            if (mutatedValue !== undefined) {
                                asyncStream.push(mutatedValue);
                                onValue?.(mutatedValue);
                            }
                        })
                            .catch((error) => {
                            onError?.(error);
                        });
                    }
                    else {
                        if (mutatedValue !== undefined) {
                            asyncStream.push(mutatedValue);
                            onValue?.(mutatedValue);
                        }
                    }
                }
                else {
                    asyncStream.push(value);
                    onValue?.(value);
                }
            }
            catch (error) {
                onError?.(error);
            }
        }
    };
    const retry = async (retries = retryAttempts) => {
        // if the stream has been retried the maximum number of times without
        // success, call onError
        if (retries === 0) {
            void asyncStream.end();
            onError?.(new StreamFailedError(retryAttempts));
            return;
        }
        // wait for the retry delay before attempting to restart the stream
        await wait(retryDelay);
        // call the onRetry callback
        onRetry?.(retryAttempts - retries + 1, retryAttempts);
        try {
            // attempt to restart the stream
            const streamCloser = await streamFunction(streamCallback, () => {
                // call the onFail callback
                onFail?.();
                void retry();
            });
            await streamCloser.waitForReady();
            // when the async stream is done, end the stream
            asyncStream.onDone = () => {
                streamCloser.end();
            };
            // stream restarted, call the onRestart callback
            onRestart?.();
        }
        catch (error) {
            onError?.(error);
            // retry
            void retry(retries - 1);
        }
    };
    const startRetry = () => {
        // if the stream should be retried, start the process
        if (retryOnFail) {
            void retry();
        }
        else {
            void asyncStream.end();
            // stream failed and should not be retried, throw an error
            onError?.(new StreamFailedError(0));
        }
    };
    try {
        // create the stream
        const streamCloser = await streamFunction(streamCallback, () => {
            // call the onFail callback
            onFail?.();
            startRetry();
        });
        await streamCloser.waitForReady();
        // when the async stream is done, end the stream
        asyncStream.onDone = () => {
            streamCloser.end();
        };
    }
    catch (error) {
        onError?.(error);
        startRetry();
    }
    // return a proxy for the async stream
    return createAsyncStreamProxy(asyncStream);
};

/**
 * Represents a conversation
 *
 * This class is not intended to be initialized directly.
 */
class Conversation {
    #client;
    #codecRegistry;
    #conversation;
    /**
     * Creates a new conversation instance
     *
     * @param client - The client instance managing the conversation
     * @param codecRegistry - The codec registry instance
     * @param conversation - The underlying conversation instance
     */
    constructor(client, codecRegistry, conversation) {
        this.#client = client;
        this.#codecRegistry = codecRegistry;
        this.#conversation = conversation;
    }
    /**
     * Gets the unique identifier for this conversation
     */
    get id() {
        return this.#conversation.id();
    }
    /**
     * Gets whether this conversation is currently active
     */
    get isActive() {
        return this.#conversation.isActive();
    }
    /**
     * Gets the inbox ID that added this client's inbox to the conversation
     */
    get addedByInboxId() {
        return this.#conversation.addedByInboxId();
    }
    /**
     * Gets the timestamp when the conversation was created in nanoseconds
     */
    get createdAtNs() {
        return this.#conversation.createdAtNs();
    }
    /**
     * Gets the date when the conversation was created
     */
    get createdAt() {
        return nsToDate(this.createdAtNs);
    }
    pausedForVersion() {
        return this.#conversation.pausedForVersion() ?? undefined;
    }
    /**
     * Gets HMAC keys for this conversation
     *
     * @returns The HMAC keys for this conversation
     */
    hmacKeys() {
        return this.#conversation.getHmacKeys();
    }
    /**
     * Gets the metadata for this conversation
     *
     * @returns Promise that resolves with the conversation metadata
     */
    async metadata() {
        const metadata = await this.#conversation.groupMetadata();
        return {
            creatorInboxId: metadata.creatorInboxId(),
            conversationType: metadata.conversationType(),
        };
    }
    /**
     * Gets the members of this conversation
     *
     * @returns Promise that resolves with the conversation members
     */
    async members() {
        return this.#conversation.listMembers();
    }
    /**
     * Synchronizes conversation data from the network
     *
     * @returns Promise that resolves when synchronization is complete
     */
    async sync() {
        return this.#conversation.sync();
    }
    /**
     * Creates a stream for new messages in this conversation
     *
     * @param options - Optional stream options
     * @returns Stream instance for new messages
     */
    async stream(options) {
        const stream = async (callback, onFail) => {
            if (!options?.disableSync) {
                await this.sync();
            }
            return this.#conversation.stream(callback, onFail);
        };
        const convertMessage = (value) => {
            const enrichedMessage = this.#client.conversations.getMessageById(value.id);
            if (enrichedMessage === undefined) {
                console.warn(`Streamed message with ID "${value.id}" not found`);
            }
            return enrichedMessage;
        };
        return createStream(stream, convertMessage, options);
    }
    /**
     * Publishes pending messages that were sent optimistically
     *
     * @returns Promise that resolves when publishing is complete
     */
    async publishMessages() {
        return this.#conversation.publishMessages();
    }
    /**
     * Sends a message with configurable delivery behavior
     *
     * @param encodedContent - The encoded content to send
     * @param sendOptions - Options for sending the message
     * @param sendOptions.shouldPush - Indicates whether this message should be
     * included in push notifications
     * @param sendOptions.isOptimistic - Indicates whether this message should be
     * sent optimistically and published later via `publishMessages`
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async send(encodedContent, sendOptions) {
        if (!encodedContent.type) {
            throw new MissingContentTypeError();
        }
        return this.#conversation.send(encodedContent, sendOptions ?? { shouldPush: false });
    }
    /**
     * Sends a text message
     *
     * @param text - The text to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendText(text, isOptimistic) {
        return this.#conversation.sendText(text, isOptimistic);
    }
    /**
     * Sends a markdown message
     *
     * @param markdown - The markdown to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendMarkdown(markdown, isOptimistic) {
        return this.#conversation.sendMarkdown(markdown, isOptimistic);
    }
    /**
     * Sends a reaction message
     *
     * @param reaction - The reaction to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendReaction(reaction, isOptimistic) {
        return this.#conversation.sendReaction(reaction, isOptimistic);
    }
    /**
     * Sends a read receipt message
     *
     * @param readReceipt - The read receipt to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendReadReceipt(isOptimistic) {
        return this.#conversation.sendReadReceipt(isOptimistic);
    }
    /**
     * Sends a reply message
     *
     * @param reply - The reply to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendReply(reply, isOptimistic) {
        return this.#conversation.sendReply(reply, isOptimistic);
    }
    /**
     * Sends a transaction reference message
     *
     * @param transactionReference - The transaction reference to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendTransactionReference(transactionReference, isOptimistic) {
        return this.#conversation.sendTransactionReference(transactionReference, isOptimistic);
    }
    /**
     * Sends a wallet send calls message
     *
     * @param walletSendCalls - The wallet send calls to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendWalletSendCalls(walletSendCalls, isOptimistic) {
        return this.#conversation.sendWalletSendCalls(walletSendCalls, isOptimistic);
    }
    /**
     * Sends a actions message
     *
     * @param actions - The actions to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendActions(actions, isOptimistic) {
        return this.#conversation.sendActions(actions, isOptimistic);
    }
    /**
     * Sends a intent message
     *
     * @param intent - The intent to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendIntent(intent, isOptimistic) {
        return this.#conversation.sendIntent(intent, isOptimistic);
    }
    /**
     * Sends an attachment message
     *
     * @param attachment - The attachment to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendAttachment(attachment, isOptimistic) {
        return this.#conversation.sendAttachment(attachment, isOptimistic);
    }
    /**
     * Sends a multi remote attachment message
     *
     * @param multiRemoteAttachment - The multi remote attachment to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendMultiRemoteAttachment(multiRemoteAttachment, isOptimistic) {
        return this.#conversation.sendMultiRemoteAttachment(multiRemoteAttachment, isOptimistic);
    }
    /**
     * Sends a remote attachment message
     *
     * @param remoteAttachment - The remote attachment to send
     * @param isOptimistic - Whether to send the message optimistically
     * @returns Promise that resolves with the message ID after it has been sent
     */
    async sendRemoteAttachment(remoteAttachment, isOptimistic) {
        return this.#conversation.sendRemoteAttachment(remoteAttachment, isOptimistic);
    }
    /**
     * Lists messages in this conversation
     *
     * @param options - Optional filtering and pagination options
     * @returns Promise that resolves with an array of decoded messages
     */
    async messages(options) {
        const messages = await this.#conversation.findEnrichedMessages(options);
        return messages.map((message) => new DecodedMessage(this.#codecRegistry, message));
    }
    /**
     * Counts messages in this conversation
     *
     * @param options - Optional filtering options
     * @returns Promise that resolves with the count of messages
     */
    async countMessages(options) {
        const count = await this.#conversation.countMessages(options);
        return count;
    }
    /**
     * Gets the last message in this conversation
     *
     * @returns Promise that resolves with the last message or undefined if none exists
     */
    async lastMessage() {
        const messages = await this.messages({
            limit: 1,
            direction: 1 /* SortDirection.Descending */,
        });
        if (messages.length > 0) {
            return messages[0];
        }
        return undefined;
    }
    /**
     * Gets the consent state for this conversation
     */
    consentState() {
        return this.#conversation.consentState();
    }
    /**
     * Updates the consent state for this conversation
     *
     * @param consentState - The new consent state to set
     */
    updateConsentState(consentState) {
        this.#conversation.updateConsentState(consentState);
    }
    /**
     * Gets the message disappearing settings for this conversation
     *
     * @returns The current message disappearing settings or undefined if not set
     */
    messageDisappearingSettings() {
        return this.#conversation.messageDisappearingSettings() ?? undefined;
    }
    /**
     * Updates message disappearing settings for this conversation
     *
     * @param fromNs - The timestamp from which messages should start disappearing
     * @param inNs - The duration after which messages should disappear
     * @returns Promise that resolves when the update is complete
     */
    async updateMessageDisappearingSettings(fromNs, inNs) {
        return this.#conversation.updateMessageDisappearingSettings({
            fromNs,
            inNs,
        });
    }
    /**
     * Removes message disappearing settings from this conversation
     *
     * @returns Promise that resolves when the settings are removed
     */
    async removeMessageDisappearingSettings() {
        return this.#conversation.removeMessageDisappearingSettings();
    }
    /**
     * Checks if message disappearing is enabled for this conversation
     *
     * @returns Whether message disappearing is enabled
     */
    isMessageDisappearingEnabled() {
        return this.#conversation.isMessageDisappearingEnabled();
    }
    /**
     * Retrieves information for this conversation to help with debugging
     *
     * @returns The debug information for this conversation
     */
    async debugInfo() {
        return this.#conversation.debugInfo();
    }
    /**
     * Retrieves the last read times for this conversation
     *
     * @returns A map keyed by inbox ID with the last read timestamp
     * (nanoseconds since epoch)
     */
    async lastReadTimes() {
        return this.#conversation.getLastReadTimes();
    }
}

/**
 * Represents a direct message conversation between two inboxes
 *
 * This class is not intended to be initialized directly.
 */
class Dm extends Conversation {
    #client;
    #codecRegistry;
    #conversation;
    /**
     * Creates a new direct message conversation instance
     *
     * @param client - The client instance managing this direct message conversation
     * @param codecRegistry - The codec registry instance
     * @param conversation - The underlying conversation instance
     */
    constructor(client, codecRegistry, conversation) {
        super(client, codecRegistry, conversation);
        this.#client = client;
        this.#codecRegistry = codecRegistry;
        this.#conversation = conversation;
    }
    /**
     * Retrieves the inbox ID of the other participant in the DM
     *
     * @returns Promise that resolves with the peer's inbox ID
     */
    get peerInboxId() {
        return this.#conversation.dmPeerInboxId();
    }
    async duplicateDms() {
        const duplicateDms = await this.#conversation.findDuplicateDms();
        return duplicateDms.map((dm) => new Dm(this.#client, this.#codecRegistry, dm));
    }
}

/**
 * Represents a group conversation between multiple inboxes
 *
 * This class is not intended to be initialized directly.
 */
class Group extends Conversation {
    #conversation;
    /**
     * Creates a new group conversation instance
     *
     * @param client - The client instance managing this group conversation
     * @param codecRegistry - The codec registry instance
     * @param conversation - The underlying conversation object
     */
    constructor(client, codecRegistry, conversation) {
        super(client, codecRegistry, conversation);
        this.#conversation = conversation;
    }
    /**
     * The name of the group
     */
    get name() {
        return this.#conversation.groupName();
    }
    /**
     * Updates the group's name
     *
     * @param name The new name for the group
     */
    async updateName(name) {
        return this.#conversation.updateGroupName(name);
    }
    /**
     * The image URL of the group
     */
    get imageUrl() {
        return this.#conversation.groupImageUrlSquare();
    }
    /**
     * Updates the group's image URL
     *
     * @param imageUrl The new image URL for the group
     */
    async updateImageUrl(imageUrl) {
        return this.#conversation.updateGroupImageUrlSquare(imageUrl);
    }
    /**
     * The description of the group
     */
    get description() {
        return this.#conversation.groupDescription();
    }
    /**
     * Updates the group's description
     *
     * @param description The new description for the group
     */
    async updateDescription(description) {
        return this.#conversation.updateGroupDescription(description);
    }
    /**
     * The app data of the group
     */
    get appData() {
        return this.#conversation.appData();
    }
    /**
     * Updates the group's app data (max 8192 bytes)
     *
     * @param appData The new app data for the group
     */
    async updateAppData(appData) {
        return this.#conversation.updateAppData(appData);
    }
    /**
     * The permissions of the group
     */
    permissions() {
        const permissions = this.#conversation.groupPermissions();
        return {
            policyType: permissions.policyType(),
            policySet: permissions.policySet(),
        };
    }
    /**
     * Updates a specific permission policy for the group
     *
     * @param permissionType The type of permission to update
     * @param policy The new permission policy
     * @param metadataField Optional metadata field for the permission
     */
    async updatePermission(permissionType, policy, metadataField) {
        return this.#conversation.updatePermissionPolicy(permissionType, policy, metadataField);
    }
    /**
     * The list of admins of the group
     */
    listAdmins() {
        return this.#conversation.adminList();
    }
    /**
     * The list of super admins of the group
     */
    listSuperAdmins() {
        return this.#conversation.superAdminList();
    }
    /**
     * Checks if an inbox is an admin of the group
     *
     * @param inboxId The inbox ID to check
     * @returns Boolean indicating if the inbox is an admin
     */
    isAdmin(inboxId) {
        return this.#conversation.isAdmin(inboxId);
    }
    /**
     * Checks if an inbox is a super admin of the group
     *
     * @param inboxId The inbox ID to check
     * @returns Boolean indicating if the inbox is a super admin
     */
    isSuperAdmin(inboxId) {
        return this.#conversation.isSuperAdmin(inboxId);
    }
    /**
     * Adds members to the group using identifiers
     *
     * @param identifiers Array of member identifiers to add
     */
    async addMembersByIdentifiers(identifiers) {
        return this.#conversation.addMembers(identifiers);
    }
    /**
     * Adds members to the group using inbox IDs
     *
     * @param inboxIds Array of inbox IDs to add
     */
    async addMembers(inboxIds) {
        return this.#conversation.addMembersByInboxId(inboxIds);
    }
    /**
     * Removes members from the group using identifiers
     *
     * @param identifiers Array of member identifiers to remove
     */
    async removeMembersByIdentifiers(identifiers) {
        return this.#conversation.removeMembers(identifiers);
    }
    /**
     * Removes members from the group using inbox IDs
     *
     * @param inboxIds Array of inbox IDs to remove
     */
    async removeMembers(inboxIds) {
        return this.#conversation.removeMembersByInboxId(inboxIds);
    }
    /**
     * Promotes a group member to admin status
     *
     * @param inboxId The inbox ID of the member to promote
     */
    async addAdmin(inboxId) {
        return this.#conversation.addAdmin(inboxId);
    }
    /**
     * Removes admin status from a group member
     *
     * @param inboxId The inbox ID of the admin to demote
     */
    async removeAdmin(inboxId) {
        return this.#conversation.removeAdmin(inboxId);
    }
    /**
     * Promotes a group member to super admin status
     *
     * @param inboxId The inbox ID of the member to promote
     */
    async addSuperAdmin(inboxId) {
        return this.#conversation.addSuperAdmin(inboxId);
    }
    /**
     * Removes super admin status from a group member
     *
     * @param inboxId The inbox ID of the super admin to demote
     */
    async removeSuperAdmin(inboxId) {
        return this.#conversation.removeSuperAdmin(inboxId);
    }
    /**
     * Request to leave the group
     */
    async requestRemoval() {
        return this.#conversation.leaveGroup();
    }
    /**
     * Checks if the current user has requested to leave the group
     *
     * @returns Boolean
     */
    isPendingRemoval() {
        return (this.#conversation.membershipState() ===
            4 /* GroupMembershipState.PendingRemove */);
    }
}

/**
 * Manages conversations
 *
 * This class is not intended to be initialized directly.
 */
class Conversations {
    #client;
    #codecRegistry;
    #conversations;
    /**
     * Creates a new conversations instance
     *
     * @param client - The client instance managing the conversations
     * @param codecRegistry - The codec registry instance
     * @param conversations - The underlying conversations instance
     */
    constructor(client, codecRegistry, conversations) {
        this.#client = client;
        this.#codecRegistry = codecRegistry;
        this.#conversations = conversations;
    }
    /**
     * Retrieves a conversation by its ID
     *
     * @param id - The conversation ID to look up
     * @returns The conversation if found, undefined otherwise
     * @see https://docs.xmtp.org/chat-apps/core-messaging/create-conversations#conversation-helper-methods
     */
    async getConversationById(id) {
        try {
            // findGroupById will throw if group is not found
            const group = this.#conversations.findGroupById(id);
            const metadata = await group.groupMetadata();
            switch (metadata.conversationType()) {
                case 1 /* ConversationType.Group */:
                    return new Group(this.#client, this.#codecRegistry, group);
                case 0 /* ConversationType.Dm */:
                    return new Dm(this.#client, this.#codecRegistry, group);
                default:
                    return undefined;
            }
        }
        catch {
            return undefined;
        }
    }
    /**
     * Retrieves a DM by inbox ID
     *
     * @param inboxId - The inbox ID to look up
     * @returns The DM if found, undefined otherwise
     * @see https://docs.xmtp.org/chat-apps/core-messaging/create-conversations#conversation-helper-methods
     */
    getDmByInboxId(inboxId) {
        try {
            // findDmByTargetInboxId will throw if group is not found
            const group = this.#conversations.findDmByTargetInboxId(inboxId);
            return new Dm(this.#client, this.#codecRegistry, group);
        }
        catch {
            return undefined;
        }
    }
    /**
     * Retrieves a DM by identifier
     *
     * @param identifier - The identifier to look up
     * @returns Promise that resolves with the DM, if found
     * @see https://docs.xmtp.org/chat-apps/core-messaging/create-conversations#conversation-helper-methods
     */
    async fetchDmByIdentifier(identifier) {
        const inboxId = await this.#client.fetchInboxIdByIdentifier(identifier);
        if (!inboxId) {
            return undefined;
        }
        return this.getDmByInboxId(inboxId);
    }
    /**
     * Retrieves a message by its ID
     *
     * @param id - The message ID to look up
     * @returns The decoded message if found, undefined otherwise
     * @see https://docs.xmtp.org/chat-apps/core-messaging/create-conversations#conversation-helper-methods
     */
    getMessageById(id) {
        try {
            // findEnrichedMessageById will throw if message is not found
            const message = this.#conversations.findEnrichedMessageById(id);
            return new DecodedMessage(this.#codecRegistry, message);
        }
        catch {
            return undefined;
        }
    }
    /**
     * Creates a new group conversation without publishing to the network
     *
     * @param options - Optional group creation options
     * @returns The new group
     * @see https://docs.xmtp.org/chat-apps/core-messaging/create-conversations#optimistically-create-a-new-group-chat
     */
    createGroupOptimistic(options) {
        const group = this.#conversations.createGroupOptimistic(options);
        return new Group(this.#client, this.#codecRegistry, group);
    }
    /**
     * Creates a new group conversation with the specified identifiers
     *
     * @param identifiers - Array of identifiers for group members
     * @param options - Optional group creation options
     * @returns The new group
     * @see https://docs.xmtp.org/chat-apps/core-messaging/create-conversations#create-a-new-group-chat
     */
    async createGroupWithIdentifiers(identifiers, options) {
        const group = await this.#conversations.createGroup(identifiers, options);
        const conversation = new Group(this.#client, this.#codecRegistry, group);
        return conversation;
    }
    /**
     * Creates a new group conversation with the specified inbox IDs
     *
     * @param inboxIds - Array of inbox IDs for group members
     * @param options - Optional group creation options
     * @returns The new group
     * @see https://docs.xmtp.org/chat-apps/core-messaging/create-conversations#create-a-new-group-chat
     */
    async createGroup(inboxIds, options) {
        const group = await this.#conversations.createGroupByInboxId(inboxIds, options);
        const conversation = new Group(this.#client, this.#codecRegistry, group);
        return conversation;
    }
    /**
     * Creates a new DM conversation with the specified identifier
     *
     * @param identifier - Identifier for the DM recipient
     * @param options - Optional DM creation options
     * @returns The new DM
     * @see https://docs.xmtp.org/agents/build-agents/create-conversations#by-ethereum-address-1
     */
    async createDmWithIdentifier(identifier, options) {
        const group = await this.#conversations.createDm(identifier, options);
        const conversation = new Dm(this.#client, this.#codecRegistry, group);
        return conversation;
    }
    /**
     * Creates a new DM conversation with the specified inbox ID
     *
     * @param inboxId - Inbox ID for the DM recipient
     * @param options - Optional DM creation options
     * @returns The new DM
     * @see https://docs.xmtp.org/agents/build-agents/create-conversations#by-inbox-id-1
     */
    async createDm(inboxId, options) {
        const group = await this.#conversations.createDmByInboxId(inboxId, options);
        const conversation = new Dm(this.#client, this.#codecRegistry, group);
        return conversation;
    }
    /**
     * Lists all conversations with optional filtering
     *
     * @param options - Optional filtering and pagination options
     * @returns Array of conversations
     * @see https://docs.xmtp.org/chat-apps/list-stream-sync/list
     */
    async list(options) {
        const groups = this.#conversations.list(options);
        const conversations = await Promise.all(groups.map(async (item) => {
            const metadata = await item.conversation.groupMetadata();
            const conversationType = metadata.conversationType();
            switch (conversationType) {
                case 0 /* ConversationType.Dm */:
                    return new Dm(this.#client, this.#codecRegistry, item.conversation);
                case 1 /* ConversationType.Group */:
                    return new Group(this.#client, this.#codecRegistry, item.conversation);
                default:
                    return undefined;
            }
        }));
        return conversations.filter((conversation) => conversation !== undefined);
    }
    /**
     * Lists all groups with optional filtering
     *
     * @param options - Optional filtering and pagination options
     * @returns Array of groups
     * @see https://docs.xmtp.org/chat-apps/list-stream-sync/list#list-existing-conversations
     */
    listGroups(options) {
        const groups = this.#conversations.list({
            ...(options ?? {}),
            conversationType: 1 /* ConversationType.Group */,
        });
        return groups.map((item) => {
            const conversation = new Group(this.#client, this.#codecRegistry, item.conversation);
            return conversation;
        });
    }
    /**
     * Lists all DMs with optional filtering
     *
     * @param options - Optional filtering and pagination options
     * @returns Array of DMs
     * @see https://docs.xmtp.org/chat-apps/list-stream-sync/list#list-existing-conversations
     */
    listDms(options) {
        const groups = this.#conversations.list({
            ...(options ?? {}),
            conversationType: 0 /* ConversationType.Dm */,
        });
        return groups.map((item) => {
            const conversation = new Dm(this.#client, this.#codecRegistry, item.conversation);
            return conversation;
        });
    }
    /**
     * Synchronizes conversations for the current client from the network
     *
     * @returns Promise that resolves when sync is complete
     * @see https://docs.xmtp.org/chat-apps/list-stream-sync/sync-and-syncall
     */
    async sync() {
        return this.#conversations.sync();
    }
    /**
     * Synchronizes all conversations and messages from the network with optional
     * consent state filtering
     *
     * @param consentStates - Optional array of consent states to filter by
     * @returns Promise that resolves when sync is complete
     * @see https://docs.xmtp.org/chat-apps/list-stream-sync/sync-and-syncall#sync-all-new-welcomes-conversations-messages-and-preferences
     */
    async syncAll(consentStates) {
        return this.#conversations.syncAllConversations(consentStates);
    }
    /**
     * Creates a stream for new conversations
     *
     * @param options - Optional stream options
     * @param options.conversationType - Optional conversation type to filter by
     * @returns Stream instance for new conversations
     * @see https://docs.xmtp.org/chat-apps/list-stream-sync/stream#stream-new-group-chat-and-dm-conversations
     */
    async stream(options) {
        const stream = async (callback, onFail) => {
            if (!options?.disableSync) {
                await this.sync();
            }
            return this.#conversations.stream(callback, onFail, options?.conversationType);
        };
        const convertConversation = async (value) => {
            const metadata = await value.groupMetadata();
            const conversationType = metadata.conversationType();
            switch (conversationType) {
                case 0 /* ConversationType.Dm */:
                    return new Dm(this.#client, this.#codecRegistry, value);
                case 1 /* ConversationType.Group */:
                    return new Group(this.#client, this.#codecRegistry, value);
                default:
                    console.warn(`Unknown conversation type: ${conversationType}`);
                    return undefined;
            }
        };
        return createStream(stream, convertConversation, options);
    }
    /**
     * Creates a stream for new group conversations
     *
     * @param options - Optional stream options
     * @returns Stream instance for new group conversations
     * @see https://docs.xmtp.org/chat-apps/list-stream-sync/stream#stream-new-group-chat-and-dm-conversations
     */
    async streamGroups(options) {
        const stream = async (callback, onFail) => {
            if (!options?.disableSync) {
                await this.sync();
            }
            return this.#conversations.stream(callback, onFail, 1 /* ConversationType.Group */);
        };
        const convertConversation = (value) => {
            return new Group(this.#client, this.#codecRegistry, value);
        };
        return createStream(stream, convertConversation, options);
    }
    /**
     * Creates a stream for new DM conversations
     *
     * @param options - Optional stream options
     * @returns Stream instance for new DM conversations
     * @see https://docs.xmtp.org/chat-apps/list-stream-sync/stream#stream-new-group-chat-and-dm-conversations
     */
    async streamDms(options) {
        const stream = async (callback, onFail) => {
            if (!options?.disableSync) {
                await this.sync();
            }
            return this.#conversations.stream(callback, onFail, 0 /* ConversationType.Dm */);
        };
        const convertConversation = (value) => {
            return new Dm(this.#client, this.#codecRegistry, value);
        };
        return createStream(stream, convertConversation, options);
    }
    /**
     * Creates a stream for all new messages
     *
     * @param options - Optional stream options
     * @param options.conversationType - Optional conversation type to filter by
     * @param options.consentStates - Optional array of consent states to filter by
     * @returns Stream instance for new messages
     * @see https://docs.xmtp.org/chat-apps/list-stream-sync/stream#stream-new-group-chat-and-dm-messages
     */
    async streamAllMessages(options) {
        const streamAllMessages = async (callback, onFail) => {
            if (!options?.disableSync) {
                await this.syncAll(options?.consentStates);
            }
            return this.#conversations.streamAllMessages(callback, onFail, options?.conversationType, options?.consentStates);
        };
        const convertMessage = (value) => {
            const enrichedMessage = this.getMessageById(value.id);
            if (enrichedMessage === undefined) {
                console.warn(`Streamed message with ID "${value.id}" not found`);
            }
            return enrichedMessage;
        };
        return createStream(streamAllMessages, convertMessage, options);
    }
    /**
     * Creates a stream for all new group messages
     *
     * @param options - Optional stream options
     * @param options.consentStates - Optional array of consent states to filter by
     * @returns Stream instance for new group messages
     * @see https://docs.xmtp.org/chat-apps/list-stream-sync/stream#stream-new-group-chat-and-dm-messages
     */
    async streamAllGroupMessages(options) {
        return this.streamAllMessages({
            ...(options ?? {}),
            conversationType: 1 /* ConversationType.Group */,
            consentStates: options?.consentStates,
        });
    }
    /**
     * Creates a stream for all new DM messages
     *
     * @param options - Optional stream options
     * @param options.consentStates - Optional array of consent states to filter by
     * @returns Stream instance for new DM messages
     * @see https://docs.xmtp.org/chat-apps/list-stream-sync/stream#stream-new-group-chat-and-dm-messages
     */
    async streamAllDmMessages(options) {
        return this.streamAllMessages({
            ...(options ?? {}),
            conversationType: 0 /* ConversationType.Dm */,
            consentStates: options?.consentStates,
        });
    }
    /**
     * Creates a stream for message deletions that streams the message IDs of
     * deleted messages
     *
     * This is a local stream, does not require network sync, and will not fail
     * like other streams.
     *
     * @param options - Optional stream options
     * @returns Stream instance for message deletions
     * @deprecated Use streamDeletedMessages instead
     */
    async streamMessageDeletions(options) {
        const stream = async (callback) => {
            return this.#conversations.streamMessageDeletions(callback);
        };
        const convertMessage = (value) => value.id;
        return createStream(stream, convertMessage, options);
    }
    /**
     * Creates a stream for message deletions that streams the deleted messages
     *
     * This is a local stream, does not require network sync, and will not fail
     * like other streams.
     *
     * @param options - Optional stream options
     * @returns Stream instance for message deletions
     */
    async streamDeletedMessages(options) {
        const stream = async (callback) => {
            return this.#conversations.streamMessageDeletions(callback);
        };
        const convertMessage = (value) => {
            return new DecodedMessage(this.#codecRegistry, value);
        };
        return createStream(stream, convertMessage, options);
    }
    /**
     * Gets the HMAC keys for all conversations
     *
     * @returns The HMAC keys for all conversations
     * @see https://docs.xmtp.org/chat-apps/push-notifs/push-notifs#get-hmac-keys-for-a-conversation
     */
    hmacKeys() {
        return this.#conversations.getHmacKeys();
    }
}

/**
 * Debug information helpers for the client
 *
 * This class is not intended to be initialized directly.
 */
class DebugInformation {
    #client;
    constructor(client) {
        this.#client = client;
    }
    apiStatistics() {
        return this.#client.apiStatistics();
    }
    apiIdentityStatistics() {
        return this.#client.apiIdentityStatistics();
    }
    apiAggregateStatistics() {
        return this.#client.apiAggregateStatistics();
    }
    clearAllStatistics() {
        this.#client.clearAllStatistics();
    }
}

/**
 * Manages user preferences and consent states
 *
 * This class is not intended to be initialized directly.
 */
class Preferences {
    #client;
    #conversations;
    /**
     * Creates a new preferences instance
     *
     * @param client - The client instance managing preferences
     * @param conversations - The underlying conversations instance
     */
    constructor(client, conversations) {
        this.#client = client;
        this.#conversations = conversations;
    }
    sync() {
        return this.#client.syncPreferences();
    }
    /**
     * Retrieves the current inbox state of this client from the local database
     *
     * @returns Promise that resolves with the inbox state
     */
    async inboxState() {
        return this.#client.inboxState(false);
    }
    /**
     * Retrieves the latest inbox state of this clientfrom the network
     *
     * @returns Promise that resolves with the inbox state
     */
    async fetchInboxState() {
        return this.#client.inboxState(true);
    }
    /**
     * Retrieves the current inbox states for specified inbox IDs from the local
     * database
     *
     * @param inboxIds - Array of inbox IDs to get state for
     * @returns Promise that resolves with the inbox states for the inbox IDs
     */
    async getInboxStates(inboxIds) {
        return this.#client.addressesFromInboxId(false, inboxIds);
    }
    /**
     * Retrieves the latest inbox states for specified inbox IDs from the network
     *
     * @param inboxIds - Array of inbox IDs to get state for
     * @returns Promise that resolves with the inbox states for the inbox IDs
     */
    async fetchInboxStates(inboxIds) {
        return this.#client.addressesFromInboxId(true, inboxIds);
    }
    /**
     * Updates consent states for multiple records
     *
     * @param consentStates - Array of consent records to update
     * @returns Promise that resolves when consent states are updated
     */
    async setConsentStates(consentStates) {
        return this.#client.setConsentStates(consentStates);
    }
    /**
     * Retrieves consent state for a specific entity
     *
     * @param entityType - Type of entity to get consent for
     * @param entity - Entity identifier
     * @returns Promise that resolves with the consent state
     */
    async getConsentState(entityType, entity) {
        return this.#client.getConsentState(entityType, entity);
    }
    /**
     * Creates a stream of consent state updates
     *
     * @param options - Optional stream options
     * @returns Stream instance for consent updates
     */
    streamConsent(options) {
        const streamConsent = async (callback, onFail) => {
            if (!options?.disableSync) {
                await this.sync();
            }
            return this.#conversations.streamConsent(callback, onFail);
        };
        return createStream(streamConsent, undefined, options);
    }
    /**
     * Creates a stream of user preference updates
     *
     * @param options - Optional stream options
     * @returns Stream instance for preference updates
     */
    streamPreferences(options) {
        const streamPreferences = async (callback, onFail) => {
            if (!options?.disableSync) {
                await this.sync();
            }
            return this.#conversations.streamPreferences(callback, onFail);
        };
        return createStream(streamPreferences, undefined, options);
    }
}

const createClient = async (identifier, options) => {
    const env = options?.env || "dev";
    const host = options?.apiUrl || ApiUrls[env];
    const gatewayHost = options?.gatewayHost || undefined;
    const isSecure = host.startsWith("https");
    const inboxId = (await getInboxIdForIdentifier(identifier, env, gatewayHost)) ||
        generateInboxId(identifier, options?.nonce);
    let dbPath;
    if (options?.dbPath === undefined) {
        // Default: auto-generated path
        dbPath = join(process.cwd(), `xmtp-${env}-${inboxId}.db3`);
    }
    else if (typeof options.dbPath === "function") {
        // Callback function: call with inbox ID
        dbPath = options.dbPath(inboxId);
    }
    else {
        // String or null: use as-is
        dbPath = options.dbPath;
    }
    const logOptions = {
        structured: options?.structuredLogging ?? false,
        level: options?.loggingLevel ?? "Off" /* LogLevel.Off */,
    };
    const historySyncUrl = options?.historySyncUrl === undefined
        ? HistorySyncUrls[env]
        : options.historySyncUrl;
    const deviceSyncWorkerMode = options?.disableDeviceSync
        ? "Disabled" /* SyncWorkerMode.Disabled */
        : "Enabled" /* SyncWorkerMode.Enabled */;
    const dbEncryptionKey = isHexString(options?.dbEncryptionKey)
        ? Buffer.from(options.dbEncryptionKey.replace(/^0x/, ""), "hex")
        : options?.dbEncryptionKey;
    return createClient$1(host, gatewayHost, isSecure, dbPath, inboxId, identifier, dbEncryptionKey, historySyncUrl, deviceSyncWorkerMode, logOptions, undefined, options?.appVersion);
};

/**
 * Client for interacting with the XMTP network
 */
class Client {
    #client;
    #codecRegistry;
    #conversations;
    #debugInformation;
    #preferences;
    #signer;
    #identifier;
    #options;
    /**
     * Creates a new XMTP client instance
     *
     * This class is not intended to be initialized directly.
     * Use `Client.create` or `Client.build` instead.
     *
     * @param options - Optional configuration for the client
     */
    constructor(options) {
        this.#options = options;
        this.#codecRegistry = new CodecRegistry([...(options?.codecs ?? [])]);
    }
    /**
     * Initializes the client with the provided identifier
     *
     * This is not meant to be called directly.
     * Use `Client.create` or `Client.build` instead.
     *
     * @param identifier - The identifier to initialize the client with
     */
    async init(identifier) {
        if (this.#client) {
            return;
        }
        this.#identifier = identifier;
        this.#client = await createClient(identifier, this.#options);
        const conversations = this.#client.conversations();
        this.#conversations = new Conversations(this, this.#codecRegistry, conversations);
        this.#debugInformation = new DebugInformation(this.#client);
        this.#preferences = new Preferences(this.#client, conversations);
    }
    /**
     * Creates a new client instance with a signer
     *
     * @param signer - The signer to use for authentication
     * @param options - Optional configuration for the client
     * @returns A new client instance
     */
    static async create(signer, options) {
        const identifier = await signer.getIdentifier();
        const client = new Client(options);
        client.#signer = signer;
        await client.init(identifier);
        if (!options?.disableAutoRegister) {
            await client.register();
        }
        return client;
    }
    /**
     * Creates a new client instance with an identifier
     *
     * Clients created with this method must already be registered.
     * Any methods called that require a signer will throw an error.
     *
     * @param identifier - The identifier to use
     * @param options - Optional configuration for the client
     * @returns A new client instance
     */
    static async build(identifier, options) {
        const client = new Client({
            ...options,
            disableAutoRegister: true,
        });
        await client.init(identifier);
        return client;
    }
    /**
     * Gets the version of libxmtp used in the bindings
     */
    get libxmtpVersion() {
        return this.#client?.libxmtpVersion();
    }
    /**
     * Gets the app version used by the client
     */
    get appVersion() {
        return this.#client?.appVersion();
    }
    /**
     * Gets the client options
     */
    get options() {
        return this.#options;
    }
    /**
     * Gets the signer associated with this client
     */
    get signer() {
        return this.#signer;
    }
    /**
     * Gets the account identifier for this client
     */
    get accountIdentifier() {
        return this.#identifier;
    }
    /**
     * Gets the inbox ID associated with this client
     */
    get inboxId() {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.inboxId();
    }
    /**
     * Gets the installation ID for this client
     */
    get installationId() {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.installationId();
    }
    /**
     * Gets the installation ID bytes for this client
     */
    get installationIdBytes() {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.installationIdBytes();
    }
    /**
     * Gets whether the client is registered with the XMTP network
     *
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    get isRegistered() {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.isRegistered();
    }
    /**
     * Gets the conversations manager for this client
     *
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    get conversations() {
        if (!this.#conversations) {
            throw new ClientNotInitializedError();
        }
        return this.#conversations;
    }
    /**
     * Gets the debug information helpersfor this client
     *
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    get debugInformation() {
        if (!this.#debugInformation) {
            throw new ClientNotInitializedError();
        }
        return this.#debugInformation;
    }
    /**
     * Gets the preferences manager for this client
     *
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    get preferences() {
        if (!this.#preferences) {
            throw new ClientNotInitializedError();
        }
        return this.#preferences;
    }
    /**
     * Adds a signature to a signature request using the client's signer (or the
     * provided signer)
     *
     * WARNING: This function should be used with caution. It is only provided
     * for use in special cases where the provided workflows do not meet the
     * requirements of an application.
     *
     * It is highly recommended to use the `register`, `unsafe_addAccount`,
     * `removeAccount`, `revokeAllOtherInstallations`, or `revokeInstallations`
     * methods instead.
     *
     * @param signatureRequest - The signature request to add the signature to
     * @throws {ClientNotInitializedError} if the client is not initialized
     * @throws {SignerUnavailableError} if no signer is available
     */
    async unsafe_addSignature(signatureRequest, signer) {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        if (!this.#signer) {
            throw new SignerUnavailableError();
        }
        const finalSigner = signer ?? this.#signer;
        const signature = await finalSigner.signMessage(await signatureRequest.signatureText());
        const identifier = await finalSigner.getIdentifier();
        switch (finalSigner.type) {
            case "SCW":
                await signatureRequest.addScwSignature(identifier, signature, finalSigner.getChainId(), finalSigner.getBlockNumber?.());
                break;
            case "EOA":
                await signatureRequest.addEcdsaSignature(signature);
                break;
        }
    }
    /**
     * Returns a signature request handler for creating a new inbox
     *
     * WARNING: This function should be used with caution. It is only provided
     * for use in special cases where the provided workflows do not meet the
     * requirements of an application.
     *
     * It is highly recommended to use the `register` method instead.
     *
     * @returns The signature text
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    async unsafe_createInboxSignatureRequest() {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.createInboxSignatureRequest();
    }
    /**
     * Returns a signature request handler for adding a new account to the
     * client's inbox
     *
     * WARNING: This function should be used with caution. It is only provided
     * for use in special cases where the provided workflows do not meet the
     * requirements of an application.
     *
     * It is highly recommended to use the `unsafe_addAccount` method instead.
     *
     * The `allowInboxReassign` parameter must be true or this function will
     * throw an error.
     *
     * @param newAccountIdentifier - The identifier of the new account
     * @param allowInboxReassign - Whether to allow inbox reassignment
     * @returns The signature text
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    async unsafe_addAccountSignatureRequest(newAccountIdentifier, allowInboxReassign = false) {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        if (!allowInboxReassign) {
            throw new InboxReassignError();
        }
        return this.#client.addIdentifierSignatureRequest(newAccountIdentifier);
    }
    /**
     * Returns a signature request handler for removing an account from the
     * client's inbox
     *
     * WARNING: This function should be used with caution. It is only provided
     * for use in special cases where the provided workflows do not meet the
     * requirements of an application.
     *
     * It is highly recommended to use the `removeAccount` method instead.
     *
     * @param identifier - The identifier of the account to remove
     * @returns The signature text
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    async unsafe_removeAccountSignatureRequest(identifier) {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.revokeIdentifierSignatureRequest(identifier);
    }
    /**
     * Returns a signature request handler for revoking all other installations
     * of the client's inbox
     *
     * WARNING: This function should be used with caution. It is only provided
     * for use in special cases where the provided workflows do not meet the
     * requirements of an application.
     *
     * It is highly recommended to use the `revokeAllOtherInstallations` method instead.
     *
     * @returns The signature text
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    async unsafe_revokeAllOtherInstallationsSignatureRequest() {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.revokeAllOtherInstallationsSignatureRequest();
    }
    /**
     * Returns a signature request handler for revoking specific installations
     * of the client's inbox
     *
     * WARNING: This function should be used with caution. It is only provided
     * for use in special cases where the provided workflows do not meet the
     * requirements of an application.
     *
     * It is highly recommended to use the `revokeInstallations` method instead.
     *
     * @param installationIds - The installation IDs to revoke
     * @returns The signature text
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    async unsafe_revokeInstallationsSignatureRequest(installationIds) {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.revokeInstallationsSignatureRequest(installationIds);
    }
    /**
     * Returns a signature request handler for changing the recovery identifier
     * for this client's inbox
     *
     * WARNING: This function should be used with caution. It is only provided
     * for use in special cases where the provided workflows do not meet the
     * requirements of an application.
     *
     * It is highly recommended to use the `changeRecoveryIdentifier` method instead.
     *
     * @param identifier - The new recovery identifier
     * @returns The signature text
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    async unsafe_changeRecoveryIdentifierSignatureRequest(identifier) {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.changeRecoveryIdentifierSignatureRequest(identifier);
    }
    /**
     * Applies a signature request to the client
     *
     * WARNING: This function should be used with caution. It is only provided
     * for use in special cases where the provided workflows do not meet the
     * requirements of an application.
     *
     * It is highly recommended to use the `register`, `unsafe_addAccount`,
     * `removeAccount`, `revokeAllOtherInstallations`, or `revokeInstallations`
     * methods instead.
     *
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    async unsafe_applySignatureRequest(signatureRequest) {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.applySignatureRequest(signatureRequest);
    }
    /**
     * Registers the client with the XMTP network
     *
     * Requires a signer, use `Client.create` to create a client with a signer.
     *
     * @throws {ClientNotInitializedError} if the client is not initialized
     * @throws {SignerUnavailableError} if no signer is available
     */
    async register() {
        const signatureRequest = await this.unsafe_createInboxSignatureRequest();
        if (!signatureRequest) {
            return;
        }
        await this.unsafe_addSignature(signatureRequest);
        await this.#client?.registerIdentity(signatureRequest);
    }
    /**
     * Adds a new account to the client inbox
     *
     * WARNING: This function should be used with caution. Adding a wallet already
     * associated with an inbox ID will cause the wallet to lose access to
     * that inbox.
     *
     * The `allowInboxReassign` parameter must be true to reassign an inbox
     * already associated with a different account.
     *
     * Requires a signer, use `Client.create` to create a client with a signer.
     *
     * @param newAccountSigner - The signer for the new account
     * @param allowInboxReassign - Whether to allow inbox reassignment
     * @throws {AccountAlreadyAssociatedError} if the account is already associated with an inbox ID
     * @throws {ClientNotInitializedError} if the client is not initialized
     * @throws {SignerUnavailableError} if no signer is available
     */
    async unsafe_addAccount(newAccountSigner, allowInboxReassign = false) {
        // check for existing inbox id
        const identifier = await newAccountSigner.getIdentifier();
        const existingInboxId = await this.fetchInboxIdByIdentifier(identifier);
        if (existingInboxId && !allowInboxReassign) {
            throw new AccountAlreadyAssociatedError(existingInboxId);
        }
        const signatureRequest = await this.unsafe_addAccountSignatureRequest(identifier, allowInboxReassign);
        await this.unsafe_addSignature(signatureRequest, newAccountSigner);
        await this.unsafe_applySignatureRequest(signatureRequest);
    }
    /**
     * Removes an account from the client's inbox
     *
     * Requires a signer, use `Client.create` to create a client with a signer.
     *
     * @param identifier - The identifier of the account to remove
     * @throws {ClientNotInitializedError} if the client is not initialized
     * @throws {SignerUnavailableError} if no signer is available
     */
    async removeAccount(identifier) {
        const signatureRequest = await this.unsafe_removeAccountSignatureRequest(identifier);
        await this.unsafe_addSignature(signatureRequest);
        await this.unsafe_applySignatureRequest(signatureRequest);
    }
    /**
     * Revokes all other installations of the client's inbox
     *
     * Requires a signer, use `Client.create` to create a client with a signer.
     *
     * @throws {ClientNotInitializedError} if the client is not initialized
     * @throws {SignerUnavailableError} if no signer is available
     */
    async revokeAllOtherInstallations() {
        const signatureRequest = await this.unsafe_revokeAllOtherInstallationsSignatureRequest();
        // no other installations to revoke
        if (!signatureRequest) {
            return;
        }
        await this.unsafe_addSignature(signatureRequest);
        await this.unsafe_applySignatureRequest(signatureRequest);
    }
    /**
     * Revokes specific installations of the client's inbox
     *
     * Requires a signer, use `Client.create` to create a client with a signer.
     *
     * @param installationIds - The installation IDs to revoke
     * @throws {ClientNotInitializedError} if the client is not initialized
     * @throws {SignerUnavailableError} if no signer is available
     */
    async revokeInstallations(installationIds) {
        const signatureRequest = await this.unsafe_revokeInstallationsSignatureRequest(installationIds);
        await this.unsafe_addSignature(signatureRequest);
        await this.unsafe_applySignatureRequest(signatureRequest);
    }
    /**
     * Revokes specific installations of the client's inbox without a client
     *
     * @param env - The environment to use
     * @param signer - The signer to use
     * @param inboxId - The inbox ID to revoke installations for
     * @param installationIds - The installation IDs to revoke
     */
    static async revokeInstallations(signer, inboxId, installationIds, env, gatewayHost) {
        const host = ApiUrls[env ?? "dev"];
        const identifier = await signer.getIdentifier();
        const signatureRequest = await revokeInstallationsSignatureRequest(host, gatewayHost, identifier, inboxId, installationIds);
        const signatureText = await signatureRequest.signatureText();
        const signature = await signer.signMessage(signatureText);
        switch (signer.type) {
            case "SCW":
                await signatureRequest.addScwSignature(identifier, signature, signer.getChainId(), signer.getBlockNumber?.());
                break;
            case "EOA":
                await signatureRequest.addEcdsaSignature(signature);
                break;
        }
        await applySignatureRequest(host, gatewayHost, signatureRequest);
    }
    /**
     * Changes the recovery identifier for the client's inbox
     *
     * Requires a signer, use `Client.create` to create a client with a signer.
     *
     * @param identifier - The new recovery identifier
     * @throws {ClientNotInitializedError} if the client is not initialized
     * @throws {SignerUnavailableError} if no signer is available
     */
    async changeRecoveryIdentifier(identifier) {
        const signatureRequest = await this.unsafe_changeRecoveryIdentifierSignatureRequest(identifier);
        await this.unsafe_addSignature(signatureRequest);
        await this.unsafe_applySignatureRequest(signatureRequest);
    }
    /**
     * Checks if the client can message the specified identifiers
     *
     * @param identifiers - The identifiers to check
     * @returns Whether the client can message the identifiers
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    async canMessage(identifiers) {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        const canMessage = await this.#client.canMessage(identifiers);
        return new Map(Object.entries(canMessage));
    }
    /**
     * Fetches the key package statuses from the network for the specified
     * installation IDs
     *
     * @param installationIds - The installation IDs to check
     * @returns The key package statuses
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    async fetchKeyPackageStatuses(installationIds) {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.getKeyPackageStatusesForInstallationIds(installationIds);
    }
    /**
     * Fetches the inbox ID for a given identifier from the local database
     * If not found, fetches from the network
     *
     * @param identifier - The identifier to look up
     * @returns The inbox ID, if found
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    async fetchInboxIdByIdentifier(identifier) {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.findInboxIdByIdentifier(identifier);
    }
    /**
     * Signs a message with the installation key
     *
     * @param signatureText - The text to sign
     * @returns The signature
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    signWithInstallationKey(signatureText) {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.signWithInstallationKey(signatureText);
    }
    /**
     * Verifies a signature was made with the installation key
     *
     * @param signatureText - The text that was signed
     * @param signatureBytes - The signature bytes to verify
     * @returns Whether the signature is valid
     * @throws {ClientNotInitializedError} if the client is not initialized
     */
    verifySignedWithInstallationKey(signatureText, signatureBytes) {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        try {
            this.#client.verifySignedWithInstallationKey(signatureText, signatureBytes);
            return true;
        }
        catch {
            return false;
        }
    }
    /**
     * Fetches the inbox states for the specified inbox IDs from the network
     * without a client
     *
     * @param env - The environment to use
     * @param inboxIds - The inbox IDs to get the state for
     * @returns The inbox states for the specified inbox IDs
     */
    static async fetchInboxStates(inboxIds, env, gatewayHost) {
        const host = ApiUrls[env ?? "dev"];
        return inboxStateFromInboxIds(host, gatewayHost, inboxIds);
    }
    /**
     * Checks if the specified identifiers can be messaged
     *
     * @param identifiers - The identifiers to check
     * @param env - Optional XMTP environment
     * @returns Map of identifiers to whether they can be messaged
     */
    static async canMessage(identifiers, env) {
        const canMessageMap = new Map();
        for (const identifier of identifiers) {
            const inboxId = await getInboxIdForIdentifier(identifier, env);
            canMessageMap.set(identifier.identifier.toLowerCase(), inboxId !== null);
        }
        return canMessageMap;
    }
    /**
     * Verifies a signature was made with a public key
     *
     * @param signatureText - The text that was signed
     * @param signatureBytes - The signature bytes to verify
     * @param publicKey - The public key to verify against
     * @returns Whether the signature is valid
     */
    static verifySignedWithPublicKey(signatureText, signatureBytes, publicKey) {
        try {
            verifySignedWithPublicKey(signatureText, signatureBytes, publicKey);
            return true;
        }
        catch {
            return false;
        }
    }
    /**
     * Checks if an address is authorized for an inbox
     *
     * @param inboxId - The inbox ID to check
     * @param address - The address to check
     * @param options - Optional network options
     * @returns Whether the address is authorized
     */
    static async isAddressAuthorized(inboxId, address, env, gatewayHost) {
        const host = ApiUrls[env ?? "dev"];
        return await isAddressAuthorized(host, gatewayHost, inboxId, address);
    }
    /**
     * Checks if an installation is authorized for an inbox
     *
     * @param inboxId - The inbox ID to check
     * @param installation - The installation to check
     * @param options - Optional network options
     * @returns Whether the installation is authorized
     */
    static async isInstallationAuthorized(inboxId, installation, env, gatewayHost) {
        const host = ApiUrls[env ?? "dev"];
        return await isInstallationAuthorized(host, gatewayHost, inboxId, installation);
    }
    /**
     * Send a sync request to other devices on the network
     *
     * @returns Promise that resolves when the sync request is sent
     */
    async sendSyncRequest() {
        if (!this.#client) {
            throw new ClientNotInitializedError();
        }
        return this.#client.sendSyncRequest();
    }
}

export { AccountAlreadyAssociatedError, ApiUrls, Client, ClientNotInitializedError, Conversation, Conversations, DecodedMessage, Dm, Group, HistorySyncUrls, InboxReassignError, MissingContentTypeError, SignerUnavailableError, StreamFailedError, StreamInvalidRetryAttemptsError, generateInboxId, getInboxIdForIdentifier, isActions, isAttachment, isGroupUpdated, isHexString, isIntent, isLeaveRequest, isMarkdown, isMultiRemoteAttachment, isReaction, isReadReceipt, isRemoteAttachment, isReply, isText, isTextReply, isTransactionReference, isWalletSendCalls, validHex };
